---
title: "CITE-seq Pilot"
output: github_document
---

```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, comment = "#")
options(width = 999)
```

## Packages

```{r}
# Data wrangling
library(tidyverse)
library(rvest)

# single-cell data analysis
library(Seurat)

# Plotting
library(ggridges)
library(RColorBrewer)
library(cowplot)
```


## Cell Ranger data

```{r}
cellranger_dir <- 
    file.path("/lab-share/IM-Gutierrez-e2/Public/scRNA/SN0231064/KW9100_Maria",
              "210726_10X_KW9100-2_bcl/cellranger-6.0.1/GRCh38/BRI-1283/outs",
              "filtered_feature_bc_matrix")

features_df <- file.path(cellranger_dir, "features.tsv.gz") %>%
    read_tsv(col_names = c("gene_id", "gene_name", "phenotype"))

mt_genes <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("^MT-", gene_name)) %>%
    pull(gene_id)

ribo_genes <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("^RPS\\d+|^RPL\\d+", gene_name))

data10x <- Read10X(cellranger_dir, gene.column = 1)
```

## Create the Seurat object

```{r}
gene_exp <- data10x[["Gene Expression"]]

antibody <- data10x[["Antibody Capture"]] %>%
    .[!grepl("^Hashtag", rownames(.)), ] 

rownames(antibody) <- rownames(antibody) %>%
    sub("_prot$", "", .) %>%
    gsub("_", ".", .)

hashtags <- data10x[["Antibody Capture"]] %>%
    .[grepl("^Hashtag", rownames(.)), ]

rownames(hashtags) <- 
    c("IgG72", "RSQ72", "IgG24", "RSQ24", "Res24", "Res00")

# Create object
bcells <- CreateSeuratObject(counts = gene_exp, project = "bcells")
bcells[["ADT"]] <- CreateAssayObject(counts = antibody)
bcells[["HTO"]] <- CreateAssayObject(counts = hashtags)

# Normalize
bcells <- NormalizeData(bcells, normalization.method = "LogNormalize")
bcells <- FindVariableFeatures(bcells, selection.method = "vst")
bcells <- ScaleData(bcells, features = VariableFeatures(bcells))

bcells <- NormalizeData(bcells, assay = "HTO", 
                        normalization.method = "CLR")

bcells <- NormalizeData(bcells, assay = "ADT", 
                        normalization.method = "CLR",
                        margin = 2)
```
## Exploring the object

```{r}
bcells
```

```{r, eval=FALSE}
#check out the output of str()
str(bcells)
```

## QC

```{r}
stims <- c("Res00", "Res24", "IgG24", "IgG72", "RSQ24", "RSQ72")

bcells[["percent_mt"]] <- 
    PercentageFeatureSet(bcells, features = mt_genes)
```

```{r, echo=FALSE}
bcells_meta <- bcells@meta.data %>%
    rownames_to_column("barcode") %>%
    select(barcode, n_reads = nCount_RNA, 
           n_genes = nFeature_RNA, percent_mt) %>%
    pivot_longer(-barcode, names_to = "feature")

bcells_meta %>%
    filter(feature %in% c("n_genes", "percent_mt")) %>%
    pivot_wider(names_from = feature, values_from = value) %>%
    ggplot(aes(percent_mt, n_genes)) +
    geom_jitter(size = .25, alpha = .25) +
    geom_density2d_filled(alpha = .5) +
    geom_density_2d(size = .25, color = "black") +
    geom_vline(xintercept = 10, linetype = 2) +
    geom_hline(yintercept = 500, linetype = 2) +
    scale_x_continuous(breaks = c(0, 5, 10, 20, 40, 60, 80)) +
    scale_y_continuous(breaks = seq(0, 1e4, by = 1e3)) +
    theme_minimal() +
    theme(panel.grid.minor = element_blank(),
          legend.position = "none") +
    labs(x = "% Mitochondrial reads", y = "# of genes")
```

## Filter out cells with high % of mitochondrial RNA

```{r}
bcells <- bcells %>%
    subset(subset = nFeature_RNA > 500 & percent_mt < 10)
```

## Demultiplex cells based on HTO

```{r}
bcells <- HTODemux(bcells, assay = "HTO", positive.quantile = 0.99)

table(bcells$HTO_classification.global)

Idents(bcells) <- "HTO_maxID"
```

```{r, echo=FALSE}
hto_data <- bcells@assays$HTO@data
hto_class_df <- tibble(barcode = colnames(hto_data),
                       maxid = bcells@meta.data$HTO_maxID)

hto_df <- hto_data %>%
    as.data.frame() %>%
    rownames_to_column("hto") %>%
    pivot_longer(-hto, names_to = "barcode") %>%
    left_join(hto_class_df, by = "barcode") %>%
    select(maxid, hto, value) %>%
    mutate(maxid = factor(maxid, levels = stims),
           hto = factor(hto, levels = stims))

ggplot(hto_df, aes(value, maxid, fill = maxid)) +
    geom_density_ridges(show.legend = FALSE, size = .25) +
    facet_wrap(~hto, nrow = 2) +
    scale_fill_brewer(palette = "Paired") +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "Expression level", y = NULL)
```


## Extract Singlets

```{r}
Idents(bcells) <- "HTO_classification.global"

bcells_singlet <- subset(bcells, idents = "Singlet")

table(bcells_singlet@meta.data$HTO_maxID)[stims]
```

## Feature quantifications

```{r, echo=FALSE, fig.width=8, fig.height=4}
bcells_singlet_meta <- bcells_singlet@meta.data %>%
    rownames_to_column("barcode") %>%
    mutate(cell_class = bcells_singlet@meta.data$HTO_maxID,
           cell_class = factor(cell_class, levels = stims)) %>%
    select(barcode, cell_class, 
           n_reads = nCount_RNA, n_genes = nFeature_RNA, percent_mt) %>%
    pivot_longer(-(barcode:cell_class), names_to = "feature")

ggplot(bcells_singlet_meta, aes(cell_class, value, fill = cell_class)) +
    geom_violin(size = .25) +
    scale_y_continuous(breaks = scales::pretty_breaks(7)) +
    scale_fill_brewer(palette = "Paired") +
    facet_wrap(~feature, nrow = 1, scales = "free") +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.title = element_blank(),
          axis.ticks.x = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid.minor = element_blank()) +
    labs(fill = NULL)
```


## PCA

```{r}
bcells_singlet <- 
    FindVariableFeatures(bcells_singlet, 
                         nfeatures = 1000,
                         selection.method = "vst")

all_genes <- rownames(bcells_singlet)

bcells_singlet <- 
    ScaleData(bcells_singlet, features = all_genes)

bcells_singlet <-
    RunPCA(bcells_singlet, features = VariableFeatures(bcells_singlet))
```

```{r, echo=FALSE, fig.height=4}
tibble(sdev = bcells_singlet@reductions$pca@stdev) %>%
    rownames_to_column("pc") %>%
    mutate(pc = fct_inorder(pc),
           var = sdev^2,
           prop_var = var/sum(var)) %>%
    filter(pc %in% 1:25) %>%
    ggplot(aes(pc, prop_var)) +
    geom_line(aes(group = 1)) +
    scale_y_continuous(labels = function(x) scales::percent(x, accuracy = 1L),
                       breaks = scales::pretty_breaks(8)) +
    theme_bw() +
    theme(panel.grid.minor = element_blank()) +
    labs(x = "Principal component", y = "Proportion of variance")
```

## UMAP

```{r, }
# Find neighboring cells
bcells_singlet <- FindNeighbors(bcells_singlet, dims = 1:20)

# Cluster
bcells_singlet <- FindClusters(bcells_singlet, resolution = 0.25)

# UMAP
bcells_singlet <- RunUMAP(bcells_singlet, dims = 1:20)
```

```{r, echo = FALSE, fig.height=8}
umap_df <- as.data.frame(bcells_singlet@reductions$umap@cell.embeddings) %>%
    as_tibble() %>%
    mutate(barcode = colnames(bcells_singlet@assays$RNA@data),
           cluster = bcells_singlet@meta.data$seurat_clusters,
           stim = bcells_singlet@meta.data$HTO_maxID,
           stim = factor(stim, levels = stims),
           percent_mt = bcells_singlet@meta.data$percent_mt)

cell_class_p <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = stim)) +
    geom_point(size = 1) +
    scale_color_brewer(palette = "Paired") +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "Hashtag classification")

seurat_cluster_p <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = cluster)) +
    geom_point(size = 1) +
    scale_color_brewer(palette = "Set1", labels = function(x) str_pad(x, 5)) +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "Seurat cluster res = 0.25")

plot_grid(cell_class_p, seurat_cluster_p, ncol = 1)
```

```{r, echo=FALSE, fig.height=4}
mki_gene_df <- features_df %>%
    filter(gene_name == "MKI67") %>%
    select(gene_id, gene_name)

mki_gene_quant <- bcells_singlet@assays$RNA@data %>% 
    .[mki_gene_df$gene_id, ,drop = FALSE] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(mki_gene_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp")

umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(mki_gene_quant, by = "barcode") %>%
    ggplot(aes(UMAP_1, UMAP_2, color = gene_exp)) +
    geom_point(size = 1) +
    scale_color_viridis_c() +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "MKI67 RNA expression", color = NULL)
```

## Downsampling


```{r}
bcells_singlet_downsamp <- subset(bcells_singlet, downsample = 194)

bcells_singlet_downsamp <- 
    FindVariableFeatures(bcells_singlet_downsamp, 
                         nfeatures = 1000,
                         selection.method = "vst")

bcells_singlet_downsamp <- 
    ScaleData(bcells_singlet_downsamp, features = VariableFeatures(bcells_singlet_downsamp))

bcells_singlet_downsamp <-
    RunPCA(bcells_singlet_downsamp, features = VariableFeatures(bcells_singlet_downsamp))
```

```{r, echo=FALSE, fig.height=4}
tibble(sdev = bcells_singlet_downsamp@reductions$pca@stdev) %>%
    rownames_to_column("pc") %>%
    mutate(pc = fct_inorder(pc),
           var = sdev^2,
           prop_var = var/sum(var)) %>%
    filter(pc %in% 1:25) %>%
    ggplot(aes(pc, prop_var)) +
    geom_line(aes(group = 1)) +
    scale_y_continuous(labels = function(x) scales::percent(x, accuracy = 1L),
                       breaks = scales::pretty_breaks(8)) +
    theme_bw() +
    theme(panel.grid.minor = element_blank()) +
    labs(x = "Principal component", y = "Proportion of variance")
```

```{r}
# Find neighboring cells
bcells_singlet_downsamp <- FindNeighbors(bcells_singlet_downsamp, dims = 1:20)

# Cluster
bcells_singlet_downsamp <- FindClusters(bcells_singlet_downsamp, resolution = 0.25)

# UMAP
bbcells_singlet_downsamp <- RunUMAP(bcells_singlet_downsamp, dims = 1:20)
```

```{r, echo=FALSE}
umap_ds_df <- as.data.frame(bcells_singlet_downsamp@reductions$umap@cell.embeddings) %>%
    as_tibble() %>%
    mutate(barcode = colnames(bcells_singlet_downsamp@assays$RNA@data),
           cluster = bcells_singlet_downsamp@meta.data$seurat_clusters,
           stim = bcells_singlet_downsamp@meta.data$HTO_maxID,
           stim = factor(stim, levels = stims))

cell_class_ds_p <- ggplot(umap_ds_df, aes(UMAP_1, UMAP_2, color = stim)) +
    geom_point(size = 1) +
    scale_color_brewer(palette = "Paired") +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "Hashtag classification")

seurat_cluster_ds_p <- ggplot(umap_ds_df, aes(UMAP_1, UMAP_2, color = cluster)) +
    geom_point(size = 1) +
    scale_color_brewer(palette = "Set1", labels = function(x) str_pad(x, 5)) +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "Seurat cluster res = 0.25")

plot_grid(cell_class_ds_p, seurat_cluster_ds_p, ncol = 1)
```



## B cell genes (RNA)

```{r, echo=FALSE, fig.height=8, fig.width=9}
bcell_genes <- 
    c("IGHD", "IGHM", "CD19", 
      "CR2", "CD27", "CD38", 
      "CD69", "CD86", "ITGAX",
      "CXCR5", "HLA-DRA", "TBX21")

bcell_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           gene_name %in% bcell_genes) %>%
    select(gene_id, gene_name)

bcell_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[bcell_genes_df$gene_id, ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(bcell_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp")

bcell_genes_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(bcell_genes_quant, by = "barcode") %>%
    group_split(gene_id) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = gene_exp)) +
            geom_point(size = .2) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5)) +
            facet_wrap(~gene_name) +
            theme_minimal() +
            theme(panel.grid = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = bcell_genes_plot_list, ncol = 3)
    
```

## B cell genes (Protein)

```{r, echo=FALSE, fig.height=8, fig.width=9}
bcell_prots <- 
    c("IgD", "IgM", "CD19", 
      "CD21", "CD27", "CD38", 
      "CD69", "CD86", "CD11c",
      "CD185.or.CXCR5", "HLA-DR")

adt_quants <- bcells_singlet@assays$ADT@data[bcell_prots, ] %>%
    as.data.frame() %>%
    rownames_to_column("ab") %>%
    as_tibble() %>%
    pivot_longer(-ab, names_to = "barcode", values_to = "ab_level")

bcell_prots_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(adt_quants, by = "barcode") %>%
    group_split(ab) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = ab_level)) +
            geom_point(size = .2) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5,
                                                         barheight = 3)) +
            facet_wrap(~ab) +
            theme_minimal() +
            theme(panel.grid = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = bcell_prots_plot_list, ncol = 3)

```

## Lupus genes


```{r, echo=FALSE, fig.height=24, fig.width=10}
sle_genes <- 
    "https://www.nature.com/articles/ng.3434/tables/2" %>%
    read_html() %>%
    html_node("table") %>%
    html_table(header = TRUE, fill = TRUE) %>%
    select(gene = `Likely causal genec`) %>%
    slice(-1) %>%
    separate_rows(gene, sep = ",") %>%
    mutate(gene = trimws(gene)) %>%
    filter(gene != "") %>%
    pull(gene)

sle_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           gene_name %in% sle_genes) %>%
    select(gene_id, gene_name)

sle_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[sle_genes_df$gene_id, ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(sle_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp") %>%
    group_by(gene_id) %>%
    filter(mean(gene_exp > 0) > .01) %>%
    ungroup()

sle_genes_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(sle_genes_quant, by = "barcode") %>%
    split(.$gene_name) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = gene_exp)) +
            geom_point(size = .2) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5)) +
            facet_wrap(~gene_name) +
            theme_minimal() +
            theme(panel.grid = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = sle_genes_plot_list, ncol = 4)

```

## TLR genes

```{r, echo=FALSE}
tlr_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("TLR", gene_name)) %>%
    select(gene_id, gene_name)

tlr_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[tlr_genes_df$gene_id, ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(tlr_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp")

umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(tlr_genes_quant, by = "barcode") %>%
    ggplot(aes(UMAP_1, UMAP_2, color = gene_exp)) +
    geom_point(size = .25) +
    scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                          guide = guide_colorbar(barwidth = .5)) +
    facet_wrap(~gene_name) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          axis.title = element_blank(),
          axis.text = element_blank()) +
    labs(color = NULL)
```

## Find marker genes for each cluster

Here, I changed the clusters to make RSQ72 a separate cluster from cluster 1, and I'm calling "IgG72-prolif" as a separate cluster from IgG72 to denote the subcluster with high MIK62 gene expression.

```{r, echo = FALSE}
cluster_renamed <- 
  tibble(cluster = Idents(bcells_singlet),
         hto = bcells_singlet@meta.data$HTO_maxID) %>%
  mutate(cluster = as.character(cluster),
         cluster = case_when(cluster == 1 & hto == "RSQ72" ~ "6",
                             TRUE ~ cluster),
         cluster = factor(cluster))

bcells_singlet_renamed <- bcells_singlet
Idents(bcells_singlet_renamed) <- cluster_renamed$cluster

new_cluster_ids <- c("IgG24", "Res24", "IgG72", "RSQ24", "Res00", "IgG72-prolif", "RSQ72")
names(new_cluster_ids) <- levels(bcells_singlet_renamed)

bcells_singlet_renamed <- RenameIdents(bcells_singlet_renamed, new_cluster_ids)

bcells_markers <- 
    FindAllMarkers(bcells_singlet_renamed, 
                   only.pos = TRUE,
                   min.pct = 1/3,
                   logfc.threshold = .5) %>%
    as_tibble()
  
top_markers <- bcells_markers %>%
  as_tibble() %>%
  left_join(select(features_df, 1:2), by = c("gene" = "gene_id")) %>%
  group_by(cluster) %>%
  top_n(10, avg_log2FC) %>%
  ungroup() %>%
  select(cluster_top = cluster, gene_name, gene_id = gene, avg_log2FC)

cluster_cell <- tibble(cluster = Idents(bcells_singlet_renamed),
                       barcode = names(Idents(bcells_singlet_renamed)))
  
cell_gene_expr <- 
  bcells_singlet_renamed@assays$RNA@data %>%
  .[unique(top_markers$gene_id), ] %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  as_tibble() %>%
  pivot_longer(-gene_id, names_to = "barcode", values_to = "logexpr")
  
sorted_cluster_ids <- c("Res00", "Res24", "IgG24", "IgG72", "IgG72-prolif", "RSQ24", "RSQ72")

top_marker_expr <- cell_gene_expr %>%
  left_join(cluster_cell, by = "barcode") %>%
  inner_join(top_markers, by = "gene_id") %>%
  mutate(cluster = factor(cluster, levels = sorted_cluster_ids)) %>%
  arrange(cluster, desc(avg_log2FC)) %>%
  mutate(gene_name = fct_inorder(gene_name))
```

## Top 10 marker genes per cluster

```{r, echo=FALSE, fig.height=7, fig.width=8}
top_marker_summary <- top_marker_expr %>%
  group_by(cluster_top, cluster, gene_name) %>%
  summarise(prop_expr = mean(logexpr > 0),
            scale_expr = mean(logexpr)) %>%
  ungroup() %>%
  mutate(cluster_top = factor(cluster_top, levels = sorted_cluster_ids)) %>%
  arrange(cluster_top, cluster, desc(prop_expr))

ggplot(top_marker_summary, aes(cluster, gene_name)) +
  geom_point(aes(size = prop_expr, color = scale_expr)) +
  geom_point(aes(size = prop_expr), color = "black", shape = 21) +
  scale_size(range = c(0.1, 4), labels = scales::percent) +
  scale_color_viridis_c(option = "magma") +
  facet_wrap(~cluster_top, scales = "free") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        panel.grid = element_blank(), 
        legend.position = c(.7, .1)) +
  labs(x = NULL, y = NULL, 
       color = "Scaled\nExpression", 
       size = "% of cells") +
  guides(color = guide_colorbar(direction = "horizontal",
                                               barheight = .5,
                                               barwidth = 10),
         size = guide_legend(direction = "horizontal",
                             keywidth = 1))
```

## Marker genes IgG vs RSQ

```{r, echo=FALSE}
bcells_markers_24 <- 
    FindMarkers(bcells_singlet_renamed, 
                   ident.1 = "IgG24",
                   ident.2 = "RSQ24",
                   only.pos = TRUE,
                   min.pct = 1/3,
                   logfc.threshold = 1) %>%
    rownames_to_column("gene") %>%
    as_tibble() %>%
    select(gene, avg_log2FC, IgG = pct.1, RSQ = pct.2)

bcells_markers_72 <- 
    FindMarkers(bcells_singlet_renamed, 
                   ident.1 = c("IgG72", "IgG72-prolif"),
                   ident.2 = "RSQ72",
                   only.pos = TRUE,
                   min.pct = 1/3,
                   logfc.threshold = 1) %>%
    rownames_to_column("gene") %>%
    as_tibble() %>%
    select(gene, avg_log2FC, IgG = pct.1, RSQ = pct.2)


markers_timep_df <- 
  bind_rows("24hr" = bcells_markers_24, "72hr" = bcells_markers_72, .id = "time") %>%
  left_join(select(features_df, 1:2), by = c("gene" = "gene_id")) %>%
  pivot_longer(IgG:RSQ, names_to = "cluster", values_to = "pct")
```


```{r, echo=FALSE, fig.height=4}
markers_timep_df %>%
  filter(time == "24hr") %>%
  arrange(desc(avg_log2FC)) %>%
  mutate(gene_name = fct_inorder(gene_name),
         tier = ntile(1:nrow(.), 2)) %>%
  ggplot(aes(cluster, gene_name)) +
  geom_point(aes(size = pct, color = avg_log2FC)) +
  geom_point(aes(size = pct), shape = 21, stroke = .25) +
  scale_color_viridis_c(option = "magma") +
  scale_size(range = c(0.1, 4), labels = scales::percent) +
  facet_wrap(~tier, scales = "free") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        strip.background = element_blank(),
        strip.text = element_blank()) +
  labs(x = NULL, y = NULL, title = "DE genes IgG vs. RSQ at 24hr",
       color = "Avg log2FC", size = "% of cells") +
  guides(color = guide_colorbar(barwidth = .6))

```


```{r, echo=FALSE, fig.height=6}
markers_timep_df %>%
  filter(time == "72hr") %>%
  arrange(desc(avg_log2FC)) %>%
  mutate(gene_name = fct_inorder(gene_name),
         tier = ntile(1:nrow(.), 3)) %>%
  ggplot(aes(cluster, gene_name)) +
  geom_point(aes(size = pct, color = avg_log2FC)) +
  geom_point(aes(size = pct), shape = 21, stroke = .25) +
  scale_color_viridis_c(option = "magma") +
  scale_size(range = c(0.1, 4), labels = scales::percent) +
  facet_wrap(~tier, scales = "free") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        strip.background = element_blank(),
        strip.text = element_blank()) +
  labs(x = NULL, y = NULL, title = "DE genes IgG vs. RSQ at 72hr",
       color = "Avg log2FC", size = "% of cells") +
  guides(color = guide_colorbar(barwidth = .6))
```



