---
title: "CITE-seq Pilots"
output: github_document
---

```{r setup, include=FALSE}
unix::rlimit_as(1e12)

library(knitr)
opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, comment = "#",
               dpi = 300)
```

```{r}
# single-cell data analysis
library(Seurat)
library(miQC)
library(scater)

# Data wrangling
library(tidyverse)

# Plotting
library(tidytext)
library(ggridges)
library(RColorBrewer)
library(scico)
library(ggsci)
library(cowplot)
```

```{r, functions}
plot_admix <- function(seurat_obj, stim_colors) {

    meta_df <- seurat_obj@meta.data %>%
	as_tibble(rownames = "barcode") %>%
	select(barcode, stim = HTO_maxID, hto_class = HTO_classification.global)
  
    hto_df <- seurat_obj@assays$HTO@counts %>%
	as_tibble(rownames = "hto") %>%
	pivot_longer(-hto, names_to = "barcode") %>%
	left_join(meta_df)

    hto_top <- hto_df %>%
	group_by(barcode, hto_class) %>%
	slice_max(n = 1, order_by = value) %>%
	select(barcode, hto_class, top_hto = hto) %>%
	ungroup()

    hto_plot_df <- hto_df %>%
	left_join(hto_top, by = c("barcode", "hto_class")) %>%
	mutate_at(vars(top_hto, stim), ~factor(., levels = names(stim_colors)))

    ggplot(hto_plot_df, aes(reorder_within(barcode, by = value, within = stim), value)) +
    geom_col(aes(fill = hto), position = "fill", width = 1.01, show.legend = FALSE) +
    scale_fill_manual(values = stim_colors) +
    facet_grid(~stim, scales = "free_x", space = "free") +
    theme_bw() +
    theme(axis.text.x = element_blank(),
	  axis.ticks.x = element_blank(),
	  axis.line = element_blank(),
	  panel.grid = element_blank(),
	  panel.border = element_blank(),
	  strip.text = element_blank()) +
    labs(x = NULL, y = NULL)
}



make_tsne_df <- function(x) {
    tsne_df <- x@reductions$tsne@cell.embeddings %>%
	as_tibble(rownames = "barcode")

    meta_df <- as_tibble(x@meta.data, rownames = "barcode") %>% 
	select(barcode, clas = HTO_classification.global, stim = HTO_maxID)

    left_join(tsne_df, meta_df, by = "barcode")
}

run_demultiplex <- function(seurat_obj, top_second_r = 0.2) {

    hto_df <- t(seurat_obj@assays$HTO@counts) %>%
	as_tibble(rownames = "barcode") %>%
	pivot_longer(-barcode, names_to = "stim") %>%
	group_by(barcode) %>%
	mutate(hto_max = stim[which.max(value)],
	       prop = value/sum(value)) %>%
	ungroup()

    top_df <- hto_df %>%
	filter(stim == hto_max) %>%
	mutate(value = log10(value+1)) %>%
	group_by(stim) %>%
	mutate(z = (value - mean(value)) / sd(value)) %>%
	ungroup() %>%
	mutate(positive = between(z, -3L, 3L)) %>%
	select(barcode, stim, positive, prop)
   
    classification_df <- hto_df %>%
	filter(stim != hto_max) %>%
	select(barcode, stim, prop) %>%
	left_join(top_df, ., by = "barcode") %>%
	group_by(barcode, stim = stim.x) %>%
	summarise(status = case_when(all(positive == TRUE) & all(prop.y < top_second_r) ~ "Singlet",
				     all(positive == TRUE) & any(prop.y >= top_second_r) ~ "Doublet",
				     all(positive == FALSE) ~ "Negative")) %>%
	ungroup()

    meta_df <- seurat_obj@meta.data %>%
	as_tibble(rownames = "barcode") %>%
	select(barcode) %>%
	left_join(classification_df, by = "barcode") %>%
	select(barcode, HTO_maxID = stim, HTO_classification.global = status) %>%
	column_to_rownames("barcode") %>%
	as.data.frame()

    seurat_obj <- AddMetaData(seurat_obj, metadata = meta_df)

    seurat_obj
}


```


```{r}
cellranger_dir_1 <- 
    file.path("/lab-share/IM-Gutierrez-e2/Public/Lab_datasets/CITEseq_pilot", 
              "SN0231064/KW9100_Maria/210726_10X_KW9100-2_bcl/cellranger-6.0.1",
              "GRCh38/BRI-1283/outs/filtered_feature_bc_matrix")

cellranger_dir_2 <- 
    file.path("/lab-share/IM-Gutierrez-e2/Public/Lab_datasets/CITEseq_pilot_2",
	      "SN0257788/broad/hptmp/curtism/bwh10x/KW10170_Maria/220617_10X_KW10170_bcl",
	      "cellranger-6.1.1/GRCh38/BRI-1743_hashing/outs/filtered_feature_bc_matrix")

features_1_df <- file.path(cellranger_dir_1, "features.tsv.gz") %>%
    read_tsv(col_names = c("gene_id", "gene_name", "phenotype"))

features_2_df <- file.path(cellranger_dir_2, "features.tsv.gz") %>%
    read_tsv(col_names = c("gene_id", "gene_name", "phenotype"))

genes_df <- features_1_df %>%
  filter(phenotype == "Gene Expression") %>%
  select(gene_id, gene_name)

mt_genes <- genes_df %>%
    filter(grepl("^MT-", gene_name)) %>%
    pull(gene_id)

ribo_genes <- genes_df %>%
    filter(grepl("^RPS\\d+|^RPL\\d+", gene_name)) %>%
    pull(gene_id)

data10x_1 <- Read10X(cellranger_dir_1, gene.column = 1)
data10x_2 <- Read10X(cellranger_dir_2, gene.column = 1)
```

```{r}
antibody_1 <- data10x_1[["Antibody Capture"]] %>%
    .[!grepl("^Hashtag", rownames(.)), ] 

antibody_2 <- data10x_2[["Antibody Capture"]] %>%
    .[!grepl("^Hashtag", rownames(.)), ] 

rownames(antibody_1) <- rownames(antibody_1) %>%
    sub("_prot$", "", .) %>%
    gsub("_", ".", .)

rownames(antibody_2) <- rownames(antibody_2) %>%
    sub("_prot$", "", .) %>%
    gsub("_", ".", .)

hashtags_1 <- data10x_1[["Antibody Capture"]] %>%
    .[grepl("^Hashtag", rownames(.)), ]

rownames(hashtags_1) <- 
    c("BCR 72hr", "TLR7 72hr", "BCR 24hr", "TLR7 24hr", "Res 24hr", "Res 0hr")

hashtags_2 <- data10x_2[["Antibody Capture"]] %>%
    .[grepl("^Hashtag", rownames(.)), ]

rownames(hashtags_2) <- 
    c("day0", 
    "IL4 24hr",
    "BCR 24hr",
    "BCR+TLR7 24hr",
    "TLR7 24hr", 
    "sCD40L 24hr",
    "CpG 24hr",
    "DN2 24hr",
    "BCR 72hr",
    "BCR+TLR7 72hr",
    "TLR7 72hr",
    "sCD40L 72hr",
    "DN2 72hr",
    "CpG 72hr")

pilot1_stim <- rownames(hashtags_1) %>%
    str_sort(numeric = TRUE) %>%
    fct_relevel("Res 0hr", after = 0) %>%
    fct_relevel("Res 24hr", after = 1) %>%
    levels() %>%
    c("Doublet", .)

pilot2_stim <- rownames(hashtags_2) %>%
    str_sort(numeric = TRUE) %>%
    fct_relevel("day0", after = 0) %>%
    fct_relevel("IL4 24hr", after = 1) %>%
    fct_relevel("TLR7 24hr", after = 4) %>%
    fct_relevel("TLR7 72hr", after = 5) %>%
    levels() %>%
    c("Doublet", .)

pilot1_pal <- c("black", "grey", "slategrey", brewer.pal(n = 4, "Paired")) %>%
    setNames(pilot1_stim)

pilot2_pal <- c("black", "grey", "slategrey", brewer.pal(n = 12, "Paired")) %>%
    setNames(pilot2_stim)


# Create object
bcells_1 <- CreateSeuratObject(counts = data10x_1[["Gene Expression"]], project = "pilot1")
bcells_1[["ADT"]] <- CreateAssayObject(counts = antibody_1)
bcells_1[["HTO"]] <- CreateAssayObject(counts = hashtags_1)

bcells_2 <- CreateSeuratObject(counts = data10x_2[["Gene Expression"]], project = "pilot2")
bcells_2[["ADT"]] <- CreateAssayObject(counts = antibody_2)
bcells_2[["HTO"]] <- CreateAssayObject(counts = hashtags_2)

bcells_1[["percent_mt"]] <- PercentageFeatureSet(bcells_1, features = mt_genes)
bcells_1[["percent_ribo"]] <- PercentageFeatureSet(bcells_1, features = ribo_genes)

bcells_2[["percent_mt"]] <- PercentageFeatureSet(bcells_2, features = mt_genes)
bcells_2[["percent_ribo"]] <- PercentageFeatureSet(bcells_2, features = ribo_genes)

# Normalize
bcells_1_m1 <- bcells_1 %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 1) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 1) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)

bcells_1_m2 <- bcells_1 %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 1) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 2) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)

bcells_2_m1 <- bcells_2 %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 1) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 1) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)

bcells_2_m2 <- bcells_2 %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 1) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 2) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)
```

```{r}
demuxlet_df <- read_tsv("./pilot_2/demuxlet/demuxlet_allsnps.best") %>%
    select(barcode = BARCODE, best = BEST) %>%
    extract(best, c("status", "sample"), "([^-]+)-(.+)")

scrublet_1 <- read_tsv("./scrublet_calls_pilot1_QC.tsv")
scrublet_2 <- read_tsv("./scrublet_calls_pilot2_QC.tsv")

names(scrublet_1)[1] <- "barcode"
names(scrublet_2)[1] <- "barcode"

```



## Raw counts

```{r, fig.width = 6, fig.height = 3}
counts_df <- 
    bind_rows(bcells_1_m1@meta.data %>%
	      rownames_to_column("barcode") %>%
	      select(barcode, pilot = orig.ident, RNA = nCount_RNA, ADT = nCount_ADT, HTO = nCount_HTO),
	      bcells_2_m1@meta.data %>%
	      rownames_to_column("barcode") %>%
	      select(barcode, pilot = orig.ident, RNA = nCount_RNA, ADT = nCount_ADT, HTO = nCount_HTO)) %>%
    as_tibble() %>%
    pivot_longer(RNA:HTO, names_to = "assay", values_to = "count")

ggplot(counts_df, aes(x = pilot, y = count)) +
    geom_violin(aes(fill = pilot), show.legend = FALSE) +
    scale_y_continuous(trans = scales::pseudo_log_trans(sigma = 1, base = 10),
		       breaks = scales::log_breaks(base = 10),
		       labels = scales::comma,
		       limits = c(10, NA)) +
    scale_fill_npg() +
    facet_wrap(~assay, nrow = 1) +
    theme_bw() +
    labs(x = NULL)
```

## Distribution of HTO counts


### Pilot 1

```{r, fig.width = 6, fig.height = 3}
hto_p1 <- as_tibble(t(bcells_1_m2@assays$HTO@counts), rownames = "barcode") %>%
    pivot_longer(-barcode, names_to = "stim") %>%
    group_by(barcode) %>%
    mutate(hto_max = stim[which.max(value)]) %>%
    ungroup() %>%
    mutate_at(vars(stim, hto_max), ~factor(., levels = names(pilot1_pal)))

ggplot(hto_p1, aes(x = log10(value+1))) +
    geom_density(aes(fill = stim), size = .25, alpha = .5) +
    scale_fill_manual(values = pilot1_pal[-1]) +
    facet_wrap(~hto_max, nrow = 2) +
    theme_bw() +
    theme(panel.grid = element_blank())
```

```{r, fig.width = 8, fig.height = 4}
hto_p2 <- as_tibble(t(bcells_2_m2@assays$HTO@counts), rownames = "barcode") %>%
    pivot_longer(-barcode, names_to = "stim") %>%
    group_by(barcode) %>%
    mutate(hto_max = stim[which.max(value)]) %>%
    ungroup() %>%
    mutate_at(vars(stim, hto_max), ~factor(., levels = names(pilot2_pal)))

ggplot(hto_p2, aes(x = log10(value+1))) +
    geom_density(aes(fill = stim), size = .25, alpha = .5) +
    scale_fill_manual(values = pilot2_pal[-1]) +
    facet_wrap(~hto_max, nrow = 3) +
    theme_bw() +
    theme(panel.grid = element_blank())
```

## QC

Here we use the miQC package to model the percentage of mitochondrial reads and number of genes, in order to identify and remove compromised cells.

```{r, fig.height=4, fig.width=4, echo = FALSE}
bcells_sce_1 <- bcells_1_m1 %>%
    as.SingleCellExperiment() %>%
    addPerCellQC(subsets = list(mito = mt_genes))

model_1 <- mixtureModel(bcells_sce_1)

bcells_sce_2 <- bcells_2_m1 %>%
    as.SingleCellExperiment() %>%
    addPerCellQC(subsets = list(mito = mt_genes))

model_2 <- mixtureModel(bcells_sce_2)

qc1 <- plotFiltering(bcells_sce_1, model_1) +
    scale_x_continuous(limits = c(0, 1e4)) +
    scale_y_continuous(limits = c(0, 100)) +
    theme_bw() +
    theme(text = element_text(size = 8),
	  panel.grid.minor = element_blank()) +
    labs(title = "Pilot #1") +
    guides(color = guide_legend(override.aes = list(size = 2)))

qc2 <- plotFiltering(bcells_sce_2, model_2) +
    scale_x_continuous(limits = c(0, 1e4)) +
    scale_y_continuous(limits = c(0, 100)) +
    theme_bw() +
    theme(text = element_text(size = 8),
	  panel.grid.minor = element_blank()) +
    labs(title = "Pilot #2") +
    guides(color = guide_legend(override.aes = list(size = 2)))

qc1$layers[[1]] <- NULL
qc2$layers[[1]] <- NULL
qc1 <- qc1 + geom_point(size = .1)
qc2 <- qc2 + geom_point(size = .1)

plot_grid(qc1, qc2, nrow = 2)
```

```{r}
bcells_sce_1_filt <- filterCells(bcells_sce_1, model_1)
bcells_sce_2_filt <- filterCells(bcells_sce_2, model_2)

cells_keep_1 <- rownames(colData(bcells_sce_1_filt))
cells_keep_2 <- rownames(colData(bcells_sce_2_filt))

bcells_1_m1_qc <- subset(bcells_1_m1, cells = cells_keep_1) %>%
  subset(nFeature_RNA > 500)

bcells_1_m2_qc <- subset(bcells_1_m2, cells = cells_keep_1) %>%
  subset(nFeature_RNA > 500)

bcells_2_m1_qc <- subset(bcells_2_m1, cells = cells_keep_2) %>%
  subset(nFeature_RNA > 500)

bcells_2_m2_qc <- subset(bcells_2_m2, cells = cells_keep_2) %>%
  subset(nFeature_RNA > 500)
```



## Demultiplex cells based on HTO

```{r, fig.width = 4, fig.height = 2.5}
bcells_1_m1_qc_demux <- HTODemux(bcells_1_m1_qc, assay = "HTO", positive.quantile = 0.99)
bcells_1_m2_qc_demux <- HTODemux(bcells_1_m2_qc, assay = "HTO", positive.quantile = 0.99)
bcells_2_m1_qc_demux <- HTODemux(bcells_2_m1_qc, assay = "HTO", positive.quantile = 0.99)
bcells_2_m2_qc_demux <- HTODemux(bcells_2_m2_qc, assay = "HTO", positive.quantile = 0.99)

class_df <-
    bind_rows("Pilot1\nmargin=1" = bcells_1_m1_qc_demux@meta.data,
	      "Pilot1\nmargin=2" = bcells_1_m2_qc_demux@meta.data,
	      "Pilot2\nmargin=1" = bcells_2_m1_qc_demux@meta.data,
	      "Pilot2\nmargin=2" = bcells_2_m2_qc_demux@meta.data,
	      .id = "id") %>%
    rownames_to_column("barcode") %>%
    as_tibble() %>%
    select(id, barcode, classification = HTO_classification.global,
	   RNA = nCount_RNA, ADT = nCount_ADT, HTO = nCount_HTO) 

class_df %>%
    count(id, classification) %>%
    ggplot(aes(id, n, fill = classification)) +
	geom_bar(position = "fill", stat = "identity", color = "black", size = .25) +
	scale_y_continuous(labels = scales::percent) +
	scale_fill_manual(values = c("Negative" = "white",
				     "Singlet" = "grey40",
				     "Doublet" = "Black")) +
	theme_bw() +
	theme(axis.text.x = element_text(size = 8)) +
	labs(x = NULL, y = "Frequency", fill = NULL)
```

```{r, fig.height = 3, fig.width = 6}
class_df2 <- class_df %>%
    select(1:3) %>%
    separate(id, c("pilot", "margin"), sep = "\n") %>%
    mutate(barcode = sub("\\.+\\d+$", "", barcode)) %>%
    group_by(pilot) %>%
    pivot_wider(names_from = margin, values_from = classification) %>%
    ungroup() %>%
    mutate_at(vars(3:4), ~factor(., levels = c("Negative", "Singlet", "Doublet")))

class_df2 %>%
    count(pilot, `margin=1`, `margin=2`) %>%
    ggplot(aes(x = `margin=1`, y = `margin=2`)) +
    geom_point(aes(fill = n, size = n),  
	       shape = 21, alpha = .5,
	       show.legend = FALSE) +
    geom_text(aes(label = n)) +
    scale_size(range = c(8, 30)) +
    scale_fill_gradient(low = "white", high = "red") +
    facet_wrap(~pilot) +
    theme_bw() +
    theme(panel.grid = element_blank())
```

```{r, fig.width=10, fig.height=3}
class_df %>%
    pivot_longer(RNA:HTO) %>%
    mutate_at(vars(classification), ~factor(., levels = c("Negative", "Singlet", "Doublet"))) %>%
    ggplot(aes(x = id, y = value)) +
    geom_violin(aes(fill = classification), size = .25) +
    facet_wrap(~name, scales = "free_y") +
    scale_y_continuous(trans = scales::pseudo_log_trans(sigma = 1, base = 10),
		       breaks = scales::log_breaks(base = 10),
		       labels = scales::comma) +
    scale_fill_manual(values = c("Negative" = "white",
				 "Singlet" = "grey40",
				 "Doublet" = "Black")) +
    theme_bw() +
    theme(axis.text.x = element_text(size = 6)) +
    labs(x = NULL, y = "Count", fill = NULL)
```

In the plot above, we see 2 differences between pilots 1 and 2:

- In pilot 1, HTO counts in higher, including for those cells being classified as "Negatives";

- In pilot 2, HTO counts for doublets not that higher than those for singlets.

Regarding the last point, doublets don't necessarily have higher counts, but a mixture of HTOs.

However, a lot of droplets being classified as doublets in pilot 2 have a profile resembling singlets in respect to the mixture of HTO, as we can see in the plots below.

```{r, fig.width = 10, fig.height = 20}
Idents(bcells_1_m1_qc_demux) <- "HTO_classification.global"
Idents(bcells_1_m2_qc_demux) <- "HTO_classification.global"
Idents(bcells_2_m1_qc_demux) <- "HTO_classification.global"
Idents(bcells_2_m2_qc_demux) <- "HTO_classification.global"


admix_1 <- plot_admix(subset(bcells_1_m1_qc_demux, idents = "Singlet"), pilot1_pal) +
    labs(title = "Pilot 1: Singlets", subtitle = "margin = 1")

admix_2 <- plot_admix(subset(bcells_1_m2_qc_demux, idents = "Singlet"), pilot1_pal) +
    labs(title = "Pilot 1: Singlets", subtitle = "margin = 2")

admix_3 <- plot_admix(subset(bcells_1_m1_qc_demux, idents = "Doublet"), pilot1_pal) +
    labs(title = "Pilot 1: Doublets", subtitle = "margin = 1")

admix_4 <- plot_admix(subset(bcells_1_m2_qc_demux, idents = "Doublet"), pilot1_pal) +
    labs(title = "Pilot 1: Doublets", subtitle = "margin = 2")

admix_5 <- plot_admix(subset(bcells_2_m1_qc_demux, idents = "Singlet"), pilot2_pal) +
    labs(title = "Pilot 2: Singlets", subtitle = "margin = 1")

admix_6 <- plot_admix(subset(bcells_2_m2_qc_demux, idents = "Singlet"), pilot2_pal) +
    labs(title = "Pilot 2: Singlets", subtitle = "margin = 2")

admix_7 <- plot_admix(subset(bcells_2_m1_qc_demux, idents = "Doublet"), pilot2_pal) +
    labs(title = "Pilot 2: Doublets", subtitle = "margin = 1")

admix_8 <- plot_admix(subset(bcells_2_m2_qc_demux, idents = "Doublet"), pilot2_pal) +
    labs(title = "Pilot 2: Doublets", subtitle = "margin = 2")

plot_grid(admix_1, admix_2, admix_3, admix_4, admix_5, admix_6, admix_7, admix_8, ncol = 1)


```{r, fig.height = 6, fig.width = 8}
bcells_1_m1_noneg <- subset(bcells_1_m1_qc_demux, idents = "Negative", invert = TRUE)
bcells_1_m2_noneg <- subset(bcells_1_m2_qc_demux, idents = "Negative", invert = TRUE)
bcells_2_m1_noneg <- subset(bcells_2_m1_qc_demux, idents = "Negative", invert = TRUE)
bcells_2_m2_noneg <- subset(bcells_2_m2_qc_demux, idents = "Negative", invert = TRUE)

DefaultAssay(bcells_1_m1_noneg) <- "HTO"
DefaultAssay(bcells_1_m2_noneg) <- "HTO"
DefaultAssay(bcells_2_m1_noneg) <- "HTO"
DefaultAssay(bcells_2_m2_noneg) <- "HTO"

bcells_1_m1_noneg <- bcells_1_m1_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunTSNE(., dims = 1:length(rownames(.)), perplexity = 100, verbose = FALSE)

bcells_1_m2_noneg <- bcells_1_m2_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunTSNE(., dims = 1:length(rownames(.)), perplexity = 100, verbose = FALSE)

bcells_2_m1_noneg <- bcells_2_m1_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunTSNE(., dims = 1:length(rownames(.)), perplexity = 100, verbose = FALSE)

bcells_2_m2_noneg <- bcells_2_m2_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunTSNE(., dims = 1:length(rownames(.)), perplexity = 100, verbose = FALSE)


tsne_dat <- bind_rows(
    "Pilot 1, Margin = 1" = make_tsne_df(bcells_1_m1_noneg),
    "Pilot 1, Margin = 2" = make_tsne_df(bcells_1_m2_noneg),
    "Pilot 2, Margin = 1" = make_tsne_df(bcells_2_m1_noneg),
    "Pilot 2, Margin = 2" = make_tsne_df(bcells_2_m2_noneg),
    .id = "set"
)


tsne_p1 <- tsne_dat %>%
    filter(grepl("Pilot 1", set)) %>%
    mutate(lab = ifelse(clas == "Doublet", clas, stim)) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = lab), size = .25) +
    scale_color_manual(values = pilot1_pal) +
    facet_wrap(~set, ncol = 2, scales = "free") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = NULL)

tsne_p2 <- tsne_dat %>%
    filter(grepl("Pilot 2", set)) %>%
    mutate(lab = ifelse(clas == "Doublet", clas, stim)) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = lab), size = .25) +
    scale_color_manual(values = pilot2_pal) +
    facet_wrap(~set, ncol = 2, scales = "free") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8),
	  legend.key.height = unit(.75, "lines")) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = NULL)

plot_grid(plot_grid(tsne_p1 + theme(legend.position = "none"), get_legend(tsne_p1), nrow = 1, rel_widths = c(1, .25)),
	  plot_grid(tsne_p2 + theme(legend.position = "none"), get_legend(tsne_p2), nrow = 1, rel_widths = c(1, .25)),
	  ncol = 1)
```

## Demultiplex by individuals' genotypes or GMM-demux in Pilot 2

```{r, fig.width = 8, fig.height = 3}
ssd_2 <- read_lines("./pilot_2/SSD_mtx/barcodes.tsv.gz")

tsne_genot <- tsne_dat %>%
    filter(set == "Pilot 2, Margin = 2") %>%
    left_join(demuxlet_df, by = "barcode") %>%
    mutate(sample = ifelse(status == "SNG", sample, status)) %>%
    select(set, barcode, tSNE_1, tSNE_2, sample) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = sample), size = .25) +
    scale_color_manual(values = c("AMB" = "salmon", "DBL" = "red", 
				  "SAMPLE1" = "cornflowerblue", "SAMPLE2" = "midnightblue")) +
    facet_wrap(~set) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = NULL)

tsne_gmm <- tsne_dat %>%
    filter(set == "Pilot 2, Margin = 2") %>%
    mutate(single_hto = barcode %in% ssd_2) %>% 
    select(set, barcode, tSNE_1, tSNE_2, single_hto) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = single_hto), size = .25, alpha = .5) +
    facet_wrap(~set) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = "Single HTO?")

plot_grid(tsne_genot, tsne_gmm, nrow = 1)

```

## Scrublet

```{r, fig.width = 8, fig.height = 3}
tsne_scr1 <- tsne_dat %>%
    filter(set == "Pilot 1, Margin = 2") %>%
    left_join(scrublet_1, by = "barcode") %>%
    select(set, barcode, tSNE_1, tSNE_2, doublet_score) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = doublet_score), size = .25) +
    scale_color_gradient(low = "white", high = "tomato4",
			 guide = guide_colorbar(barwidth = .5)) +
    facet_wrap(~set) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    labs(color = "Doublet\n score")

tsne_scr2 <- tsne_dat %>%
    filter(set == "Pilot 2, Margin = 2") %>%
    left_join(scrublet_2, by = "barcode") %>%
    select(set, barcode, tSNE_1, tSNE_2, doublet_score) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = doublet_score), size = .25) +
    scale_color_gradient(low = "white", high = "tomato4",
			 guide = guide_colorbar(barwidth = .5)) +
    facet_wrap(~set) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    labs(color = "Doublet\n score")

plot_grid(tsne_scr1, tsne_scr2, nrow = 1)

```

## Custom demultiplexing based on HTO

```{r}
bcells_1_qc_customdx <-  run_demultiplex(bcells_1_m1_qc)
bcells_2_qc_customdx <-  run_demultiplex(bcells_2_m1_qc)
```

## Number of cells for custom vs Seurat demultiplex method

```{r, fig.width = 8, fig.height = 2.5}
class_tmp_1 <- class_df %>% 
    count(id, classification) %>%
    separate(id, c("pilot", "id"), sep = "\n") %>%
    mutate(id = paste0("Seurat-", id))

class_tmp_2 <- 
    bind_rows("Pilot1-custom" = count(bcells_1_qc_customdx@meta.data, HTO_classification.global),
	      "Pilot2-custom" = count(bcells_2_qc_customdx@meta.data, HTO_classification.global),
	      .id = "id") %>%
    separate(id, c("pilot", "id"), sep = "-") %>%
    select(pilot, id, classification = 3, n)

bind_rows(class_tmp_1, class_tmp_2) %>%
    mutate(classification = factor(classification, levels = c("Negative", "Singlet", "Doublet"))) %>%
    ggplot(aes(id, n, fill = classification)) +
    geom_col(position = "dodge", color = "black") +
    scale_fill_manual(values = c("Negative" = "white", "Singlet" = "grey", "Doublet" = "black")) +
    scale_x_discrete(labels = function(x) sub("-", "\n", x)) +
    facet_wrap(~pilot, nrow = 1) +
    theme_minimal() +
    theme(panel.grid.major.x = element_blank()) +
    labs(x = NULL, y = "Number of droplets")
```

## Profile of Singlets and Doublets in custom demultiplexing

```{r, fig.width = 10, fig.height = 10}
Idents(bcells_1_qc_customdx) <- "HTO_classification.global"
Idents(bcells_2_qc_customdx) <- "HTO_classification.global"

admix_custom_1 <- plot_admix(subset(bcells_1_qc_customdx, idents = "Singlet"), pilot1_pal) +
    labs(title = "Pilot 1: Singlets")

admix_custom_2 <- plot_admix(subset(bcells_1_qc_customdx, idents = "Doublet"), pilot1_pal) +
    labs(title = "Pilot 1: Doublets")

admix_custom_3 <- plot_admix(subset(bcells_2_qc_customdx, idents = "Singlet"), pilot2_pal) +
    labs(title = "Pilot 2: Singlets")

admix_custom_4 <- plot_admix(subset(bcells_2_qc_customdx, idents = "Doublet"), pilot2_pal) +
    labs(title = "Pilot 2: Doublets")

plot_grid(admix_custom_1, admix_custom_2, admix_custom_3, admix_custom_4, ncol = 1)
```

```{r}
bcells_1_qc_customdx_noneg <- subset(bcells_1_qc_customdx, idents = "Negative", invert = TRUE)
bcells_2_qc_customdx_noneg <- subset(bcells_2_qc_customdx, idents = "Negative", invert = TRUE)

DefaultAssay(bcells_1_qc_customdx_noneg) <- "HTO"
DefaultAssay(bcells_2_qc_customdx_noneg) <- "HTO"

bcells_1_qc_customdx_noneg <- bcells_1_qc_customdx_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunTSNE(., dims = 1:length(rownames(.)), perplexity = 100, verbose = FALSE)

bcells_2_qc_customdx_noneg <- bcells_2_qc_customdx_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunTSNE(., dims = 1:length(rownames(.)), perplexity = 100, verbose = FALSE)

tsne_dat_custom <- bind_rows(
    "Pilot 1" = make_tsne_df(bcells_1_qc_customdx_noneg),
    "Pilot 2" = make_tsne_df(bcells_2_qc_customdx_noneg),
    .id = "set"
)
```

```{r, fig.width = 8, fig.height = 3}
tsne_custom_p1 <- tsne_dat_custom %>%
    filter(set == "Pilot 1") %>%
    mutate(lab = ifelse(clas == "Doublet", clas, stim)) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = lab), size = .25) +
    scale_color_manual(values = pilot1_pal) +
    facet_wrap(~set, ncol = 2, scales = "free") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = NULL)

tsne_custom_p2 <- tsne_dat_custom %>%
    filter(grepl("Pilot 2", set)) %>%
    mutate(lab = ifelse(clas == "Doublet", clas, stim)) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = lab), size = .25) +
    scale_color_manual(values = pilot2_pal) +
    facet_wrap(~set, ncol = 2, scales = "free") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8),
	  legend.key.height = unit(.75, "lines")) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = NULL)

plot_grid(plot_grid(tsne_custom_p1 + theme(legend.position = "none"), 
		    get_legend(tsne_custom_p1), nrow = 1, rel_widths = c(1, .25)),
	  plot_grid(tsne_custom_p2 + theme(legend.position = "none"), 
		    get_legend(tsne_custom_p2), nrow = 1, rel_widths = c(1, .25)),
	  nrow = 1)
```

```{r, fig.width = 8, fig.height = 3}
tsne_custom_scr1 <- tsne_dat_custom %>%
    filter(set == "Pilot 1") %>%
    left_join(scrublet_1, by = "barcode") %>%
    select(set, barcode, tSNE_1, tSNE_2, doublet_score) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = doublet_score), size = .25) +
    scale_color_gradient(low = "white", high = "tomato4",
			 guide = guide_colorbar(barwidth = .5)) +
    facet_wrap(~set) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    labs(color = "Doublet\n score")

tsne_custom_scr2 <- tsne_dat_custom %>%
    filter(set == "Pilot 2") %>%
    left_join(scrublet_2, by = "barcode") %>%
    select(set, barcode, tSNE_1, tSNE_2, doublet_score) %>%
    ggplot(aes(tSNE_1, tSNE_2)) +
    geom_point(aes(color = doublet_score), size = .25) +
    scale_color_gradient(low = "white", high = "tomato4",
			 guide = guide_colorbar(barwidth = .5)) +
    facet_wrap(~set) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    labs(color = "Doublet\n score")

plot_grid(tsne_custom_scr1, tsne_custom_scr2, nrow = 1)

```

```{r, include=FALSE, eval=TRUE}
finalcells_1 <- bcells_1_qc_customdx@meta.data %>%
    as_tibble(rownames = "barcode") %>%
    left_join(scrublet_1, by = "barcode") %>%
    filter(HTO_classification.global == "Singlet", predicted_doublets == FALSE) %>%
    pull(barcode)

finalcells_2 <- bcells_2_qc_customdx@meta.data %>%
    as_tibble(rownames = "barcode") %>%
    left_join(scrublet_2, by = "barcode") %>%
    left_join(demuxlet_df, by = "barcode") %>%
    filter(HTO_classification.global == "Singlet", predicted_doublets == FALSE, status == "SNG") %>%
    pull(barcode)

bcells_1_final <- subset(bcells_1_qc_customdx, cells = finalcells_1)
bcells_2_final <- subset(bcells_2_qc_customdx, cells = finalcells_2)

saveRDS(bcells_1_final, file = "data/pilot1.Rdata")
saveRDS(bcells_2_final, file = "data/pilot2.Rdata")
```

