---
title: "CITE-seq Pilots"
output: github_document
---

```{r setup, include=FALSE}
unix::rlimit_as(1e12)

library(knitr)
opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, comment = "#",
               dpi = 300)
options(width = 999)
```

```{r, echo=FALSE}
# single-cell data analysis
library(Seurat)
library(miQC)
library(scater)

# Data wrangling
library(tidyverse)

# Plotting
library(tidytext)
library(ggridges)
library(RColorBrewer)
library(scico)
library(ggsci)
library(cowplot)
```


```{r, echo = FALSE}
cellranger_dir_1 <- 
    file.path("/lab-share/IM-Gutierrez-e2/Public/Lab_datasets/CITEseq_pilot", 
              "SN0231064/KW9100_Maria/210726_10X_KW9100-2_bcl/cellranger-6.0.1",
              "GRCh38/BRI-1283/outs/filtered_feature_bc_matrix")

cellranger_dir_2 <- 
    file.path("/lab-share/IM-Gutierrez-e2/Public/Lab_datasets/CITEseq_pilot_2",
	      "SN0257788/broad/hptmp/curtism/bwh10x/KW10170_Maria/220617_10X_KW10170_bcl",
	      "cellranger-6.1.1/GRCh38/BRI-1743_hashing/outs/filtered_feature_bc_matrix")

features_1_df <- file.path(cellranger_dir_1, "features.tsv.gz") %>%
    read_tsv(col_names = c("gene_id", "gene_name", "phenotype"))

features_2_df <- file.path(cellranger_dir_2, "features.tsv.gz") %>%
    read_tsv(col_names = c("gene_id", "gene_name", "phenotype"))

genes_df <- features_1_df %>%
  filter(phenotype == "Gene Expression") %>%
  select(gene_id, gene_name)

mt_genes <- genes_df %>%
    filter(grepl("^MT-", gene_name)) %>%
    pull(gene_id)

ribo_genes <- genes_df %>%
    filter(grepl("^RPS\\d+|^RPL\\d+", gene_name)) %>%
    pull(gene_id)

data10x_1 <- Read10X(cellranger_dir_1, gene.column = 1)
data10x_2 <- Read10X(cellranger_dir_2, gene.column = 1)
```

```{r, echo = FALSE}
antibody_1 <- data10x_1[["Antibody Capture"]] %>%
    .[!grepl("^Hashtag", rownames(.)), ] 

antibody_2 <- data10x_2[["Antibody Capture"]] %>%
    .[!grepl("^Hashtag", rownames(.)), ] 

rownames(antibody_1) <- rownames(antibody_1) %>%
    sub("_prot$", "", .) %>%
    gsub("_", ".", .)

rownames(antibody_2) <- rownames(antibody_2) %>%
    sub("_prot$", "", .) %>%
    gsub("_", ".", .)

hashtags_1 <- data10x_1[["Antibody Capture"]] %>%
    .[grepl("^Hashtag", rownames(.)), ]

rownames(hashtags_1) <- 
    c("BCR 72hr", "TLR7 72hr", "BCR 24hr", "TLR7 24hr", "Res 24hr", "Res 0hr")

hashtags_2 <- data10x_2[["Antibody Capture"]] %>%
    .[grepl("^Hashtag", rownames(.)), ]

rownames(hashtags_2) <- 
    c("day0", 
    "IL4 24hr",
    "BCR 24hr",
    "BCR+TLR7 24hr",
    "TLR7 24hr", 
    "sCD40L 24hr",
    "CpG 24hr",
    "DN2 24hr",
    "BCR 72hr",
    "BCR+TLR7 72hr",
    "TLR7 72hr",
    "sCD40L 72hr",
    "DN2 72hr",
    "CpG 72hr")

# Create object
bcells_1 <- CreateSeuratObject(counts = data10x_1[["Gene Expression"]], project = "pilot1")
bcells_1[["ADT"]] <- CreateAssayObject(counts = antibody_1)
bcells_1[["HTO"]] <- CreateAssayObject(counts = hashtags_1)

bcells_2 <- CreateSeuratObject(counts = data10x_2[["Gene Expression"]], project = "pilot2")
bcells_2[["ADT"]] <- CreateAssayObject(counts = antibody_2)
bcells_2[["HTO"]] <- CreateAssayObject(counts = hashtags_2)

bcells_1[["percent_mt"]] <- PercentageFeatureSet(bcells_1, features = mt_genes)
bcells_1[["percent_ribo"]] <- PercentageFeatureSet(bcells_1, features = ribo_genes)

bcells_2[["percent_mt"]] <- PercentageFeatureSet(bcells_2, features = mt_genes)
bcells_2[["percent_ribo"]] <- PercentageFeatureSet(bcells_2, features = ribo_genes)

# Normalize
bcells_1_m1 <- bcells_1 %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 1) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 1) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)

bcells_1_m2 <- bcells_1 %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 1) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 2) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)

bcells_2_m1 <- bcells_2 %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 1) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 1) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)

bcells_2_m2 <- bcells_2 %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 1) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 2) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)

```

## Raw counts

```{r, echo = FALSE, fig.width = 6, fig.height = 3}
counts_df <- 
    bind_rows(bcells_1_m1@meta.data %>%
	      rownames_to_column("barcode") %>%
	      select(barcode, pilot = orig.ident, RNA = nCount_RNA, ADT = nCount_ADT, HTO = nCount_HTO),
	      bcells_2_m1@meta.data %>%
	      rownames_to_column("barcode") %>%
	      select(barcode, pilot = orig.ident, RNA = nCount_RNA, ADT = nCount_ADT, HTO = nCount_HTO)) %>%
    as_tibble() %>%
    pivot_longer(RNA:HTO, names_to = "assay", values_to = "count")

ggplot(counts_df, aes(x = pilot, y = count)) +
    geom_violin(aes(fill = pilot), show.legend = FALSE) +
    scale_y_continuous(trans = scales::pseudo_log_trans(sigma = 1, base = 10),
		       breaks = scales::log_breaks(base = 10),
		       labels = scales::comma,
		       limits = c(10, NA)) +
    scale_fill_npg() +
    facet_wrap(~assay, nrow = 1) +
    theme_bw() +
    labs(x = NULL)

```


## QC

Here we use the miQC package to model the percentage of mitochondrial reads and number of genes, in order to identify and remove compromised cells.

```{r, echo=FALSE, fig.height=4, fig.width=4, echo = FALSE}
bcells_sce_1 <- bcells_1_m1 %>%
    as.SingleCellExperiment() %>%
    addPerCellQC(subsets = list(mito = mt_genes))

model_1 <- mixtureModel(bcells_sce_1)

bcells_sce_2 <- bcells_2_m1 %>%
    as.SingleCellExperiment() %>%
    addPerCellQC(subsets = list(mito = mt_genes))

model_2 <- mixtureModel(bcells_sce_2)

qc1 <- plotFiltering(bcells_sce_1, model_1) +
    scale_x_continuous(limits = c(0, 1e4)) +
    scale_y_continuous(limits = c(0, 100)) +
    theme_bw() +
    theme(text = element_text(size = 8),
	  panel.grid.minor = element_blank()) +
    labs(title = "Pilot #1") +
    guides(color = guide_legend(override.aes = list(size = 2)))

qc2 <- plotFiltering(bcells_sce_2, model_2) +
    scale_x_continuous(limits = c(0, 1e4)) +
    scale_y_continuous(limits = c(0, 100)) +
    theme_bw() +
    theme(text = element_text(size = 8),
	  panel.grid.minor = element_blank()) +
    labs(title = "Pilot #2") +
    guides(color = guide_legend(override.aes = list(size = 2)))

qc1$layers[[1]] <- NULL
qc2$layers[[1]] <- NULL
qc1 <- qc1 + geom_point(size = .1)
qc2 <- qc2 + geom_point(size = .1)

plot_grid(qc1, qc2, nrow = 2)
```

```{r, echo = FALSE}
bcells_sce_1_filt <- filterCells(bcells_sce_1, model_1)
bcells_sce_2_filt <- filterCells(bcells_sce_2, model_2)

cells_keep_1 <- rownames(colData(bcells_sce_1_filt))
cells_keep_2 <- rownames(colData(bcells_sce_2_filt))

bcells_1_m1_qc <- subset(bcells_1_m1, cells = cells_keep_1) %>%
  subset(nFeature_RNA > 500)

bcells_1_m2_qc <- subset(bcells_1_m2, cells = cells_keep_1) %>%
  subset(nFeature_RNA > 500)

bcells_2_m1_qc <- subset(bcells_2_m1, cells = cells_keep_2) %>%
  subset(nFeature_RNA > 500)

bcells_2_m2_qc <- subset(bcells_2_m2, cells = cells_keep_2) %>%
  subset(nFeature_RNA > 500)
```

## Demultiplex cells based on HTO

```{r, echo = FALSE, fig.width = 4, fig.height = 2.5}
bcells_1_m1_qc_demux <- HTODemux(bcells_1_m1_qc, assay = "HTO", positive.quantile = 0.99)
bcells_1_m2_qc_demux <- HTODemux(bcells_1_m2_qc, assay = "HTO", positive.quantile = 0.99)
bcells_2_m1_qc_demux <- HTODemux(bcells_2_m1_qc, assay = "HTO", positive.quantile = 0.99)
bcells_2_m2_qc_demux <- HTODemux(bcells_2_m2_qc, assay = "HTO", positive.quantile = 0.99)

class_df <-
    bind_rows("Pilot1\nmargin=1" = bcells_1_m1_qc_demux@meta.data,
	      "Pilot1\nmargin=2" = bcells_1_m2_qc_demux@meta.data,
	      "Pilot2\nmargin=1" = bcells_2_m1_qc_demux@meta.data,
	      "Pilot2\nmargin=2" = bcells_2_m2_qc_demux@meta.data,
	      .id = "id") %>%
    rownames_to_column("barcode") %>%
    as_tibble() %>%
    select(id, barcode, classification = HTO_classification.global,
	   RNA = nCount_RNA, ADT = nCount_ADT, HTO = nCount_HTO) 

class_df %>%
    count(id, classification) %>%
    ggplot(aes(id, n, fill = classification)) +
	geom_bar(position = "fill", stat = "identity", color = "black", size = .25) +
	scale_y_continuous(labels = scales::percent) +
	scale_fill_manual(values = c("Negative" = "white",
				     "Singlet" = "grey40",
				     "Doublet" = "Black")) +
	theme_bw() +
	theme(axis.text.x = element_text(size = 8)) +
	labs(x = NULL, y = "Frequency", fill = NULL)
```

```{r, echo=FALSE, fig.height = 3, fig.width = 6}
class_df2 <- class_df %>%
    select(1:3) %>%
    separate(id, c("pilot", "margin"), sep = "\n") %>%
    mutate(barcode = sub("\\.+\\d+$", "", barcode)) %>%
    group_by(pilot) %>%
    pivot_wider(names_from = margin, values_from = classification) %>%
    ungroup() %>%
    mutate_at(vars(3:4), ~factor(., levels = c("Negative", "Singlet", "Doublet")))

class_df2 %>%
    count(pilot, `margin=1`, `margin=2`) %>%
    ggplot(aes(x = `margin=1`, y = `margin=2`)) +
    geom_point(aes(fill = n, size = n),  
	       shape = 21, alpha = .5,
	       show.legend = FALSE) +
    geom_text(aes(label = n)) +
    scale_size(range = c(8, 30)) +
    scale_fill_gradient(low = "white", high = "red") +
    facet_wrap(~pilot) +
    theme_bw() +
    theme(panel.grid = element_blank())
```



```{r, echo=FALSE, fig.width=10, fig.height=3}
class_df %>%
    pivot_longer(RNA:HTO) %>%
    mutate_at(vars(classification), ~factor(., levels = c("Negative", "Singlet", "Doublet"))) %>%
    ggplot(aes(x = id, y = value)) +
    geom_violin(aes(fill = classification), size = .25) +
    facet_wrap(~name, scales = "free_y") +
    scale_y_continuous(trans = scales::pseudo_log_trans(sigma = 1, base = 10),
		       breaks = scales::log_breaks(base = 10),
		       labels = scales::comma) +
    scale_fill_manual(values = c("Negative" = "white",
				 "Singlet" = "grey40",
				 "Doublet" = "Black")) +
    theme_bw() +
    theme(axis.text.x = element_text(size = 6)) +
    labs(x = NULL, y = "Count", fill = NULL)
```

```{r, echo=FALSE, fig.height = 6, fig.width = 8}
Idents(bcells_1_m1_qc_demux) <- "HTO_classification.global"
Idents(bcells_1_m2_qc_demux) <- "HTO_classification.global"
Idents(bcells_2_m1_qc_demux) <- "HTO_classification.global"
Idents(bcells_2_m2_qc_demux) <- "HTO_classification.global"

bcells_1_m1_noneg <- subset(bcells_1_m1_qc_demux, idents = "Negative", invert = TRUE)
bcells_1_m2_noneg <- subset(bcells_1_m2_qc_demux, idents = "Negative", invert = TRUE)
bcells_2_m1_noneg <- subset(bcells_2_m1_qc_demux, idents = "Negative", invert = TRUE)
bcells_2_m2_noneg <- subset(bcells_2_m2_qc_demux, idents = "Negative", invert = TRUE)

DefaultAssay(bcells_1_m1_noneg) <- "HTO"
DefaultAssay(bcells_1_m2_noneg) <- "HTO"
DefaultAssay(bcells_2_m1_noneg) <- "HTO"
DefaultAssay(bcells_2_m2_noneg) <- "HTO"

bcells_1_m1_noneg <- bcells_1_m1_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunUMAP(., dims = 1:length(rownames(.)), verbose = FALSE)

bcells_1_m2_noneg <- bcells_1_m2_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunUMAP(., dims = 1:length(rownames(.)), verbose = FALSE)

bcells_2_m1_noneg <- bcells_2_m1_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunUMAP(., dims = 1:length(rownames(.)), verbose = FALSE)

bcells_2_m2_noneg <- bcells_2_m2_noneg %>%
    ScaleData(., features = rownames(.), verbose = FALSE) %>%
    RunPCA(., features = rownames(.), approx = FALSE, verbose = FALSE) %>%
    RunUMAP(., dims = 1:length(rownames(.)), verbose = FALSE)


make_umap_df <- function(x) {
    umap_df <- x@reductions$umap@cell.embeddings %>%
	as_tibble(rownames = "barcode")

    meta_df <- as_tibble(x@meta.data, rownames = "barcode") %>% 
	select(barcode, clas = HTO_classification.global, stim = HTO_maxID)

    left_join(umap_df, meta_df, by = "barcode")
}

umap_dat <- bind_rows(
    "Pilot 1, Margin = 1" = make_umap_df(bcells_1_m1_noneg),
    "Pilot 1, Margin = 2" = make_umap_df(bcells_1_m2_noneg),
    "Pilot 2, Margin = 1" = make_umap_df(bcells_2_m1_noneg),
    "Pilot 2, Margin = 2" = make_umap_df(bcells_2_m2_noneg),
    .id = "set"
)


pilot1_stim <- rownames(hashtags_1) %>%
    str_sort(numeric = TRUE) %>%
    fct_relevel("Res 0hr", after = 0) %>%
    fct_relevel("Res 24hr", after = 1) %>%
    levels() %>%
    c("Doublet", .)

pilot2_stim <- rownames(hashtags_2) %>%
    str_sort(numeric = TRUE) %>%
    fct_relevel("day0", after = 0) %>%
    fct_relevel("IL4 24hr", after = 1) %>%
    fct_relevel("TLR7 24hr", after = 4) %>%
    fct_relevel("TLR7 72hr", after = 5) %>%
    levels() %>%
    c("Doublet", .)

pilot1_pal <- c("black", "grey", "slategrey", brewer.pal(n = 4, "Paired")) %>%
    setNames(pilot1_stim)

pilot2_pal <- c("black", "grey", "slategrey", brewer.pal(n = 12, "Paired")) %>%
    setNames(pilot2_stim)

umap_p1 <- umap_dat %>%
    filter(grepl("Pilot 1", set)) %>%
    mutate(lab = ifelse(clas == "Doublet", clas, stim)) %>%
    ggplot(aes(UMAP_1, UMAP_2)) +
    geom_point(aes(color = lab), size = .25) +
    scale_color_manual(values = pilot1_pal) +
    facet_wrap(~set, ncol = 2, scales = "free") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8)) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = NULL)

umap_p2 <- umap_dat %>%
    filter(grepl("Pilot 2", set)) %>%
    mutate(lab = ifelse(clas == "Doublet", clas, stim)) %>%
    ggplot(aes(UMAP_1, UMAP_2)) +
    geom_point(aes(color = lab), size = .25) +
    scale_color_manual(values = pilot2_pal) +
    facet_wrap(~set, ncol = 2, scales = "free") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
	  axis.title = element_text(size = 8),
	  legend.key.height = unit(.75, "lines")) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = NULL)

plot_grid(plot_grid(umap_p1 + theme(legend.position = "none"), get_legend(umap_p1), nrow = 1, rel_widths = c(1, .25)),
	  plot_grid(umap_p2 + theme(legend.position = "none"), get_legend(umap_p2), nrow = 1, rel_widths = c(1, .25)),
	  ncol = 1)
```

