---
title: "CITE-seq Pilot"
output: github_document
---

```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, comment = "#",
               dpi = 300)
options(width = 999)
```

## Packages

```{r}
# single-cell data analysis
library(Seurat)
library(miQC)
library(scater)

# Data wrangling
library(tidyverse)
library(rvest)

# Plotting
library(tidytext)
library(ggridges)
library(RColorBrewer)
library(cowplot)
```


## Cell Ranger data

```{r, cache=TRUE}
cellranger_dir <- 
    file.path("/lab-share/IM-Gutierrez-e2/Public/scRNA/SN0231064/KW9100_Maria",
              "210726_10X_KW9100-2_bcl/cellranger-6.0.1/GRCh38/BRI-1283/outs",
              "filtered_feature_bc_matrix")

features_df <- file.path(cellranger_dir, "features.tsv.gz") %>%
    read_tsv(col_names = c("gene_id", "gene_name", "phenotype"))

mt_genes <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("^MT-", gene_name)) %>%
    pull(gene_id)

ribo_genes <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("^RPS\\d+|^RPL\\d+", gene_name)) %>%
  pull(gene_id)

data10x <- Read10X(cellranger_dir, gene.column = 1)
```

## Create the Seurat object

```{r, cache=TRUE}
gene_exp <- data10x[["Gene Expression"]]

antibody <- data10x[["Antibody Capture"]] %>%
    .[!grepl("^Hashtag", rownames(.)), ] 

rownames(antibody) <- rownames(antibody) %>%
    sub("_prot$", "", .) %>%
    gsub("_", ".", .)

hashtags <- data10x[["Antibody Capture"]] %>%
    .[grepl("^Hashtag", rownames(.)), ]

rownames(hashtags) <- 
    c("IgG72", "RSQ72", "IgG24", "RSQ24", "Res24", "Res00")

# Create object
bcells <- CreateSeuratObject(counts = gene_exp, project = "bcells")
bcells[["ADT"]] <- CreateAssayObject(counts = antibody)
bcells[["HTO"]] <- CreateAssayObject(counts = hashtags)

# Normalize
bcells <- NormalizeData(bcells, normalization.method = "LogNormalize")
bcells <- FindVariableFeatures(bcells, selection.method = "vst")
bcells <- ScaleData(bcells, features = VariableFeatures(bcells))

bcells <- NormalizeData(bcells, assay = "HTO", 
                        normalization.method = "CLR")

bcells <- NormalizeData(bcells, assay = "ADT", 
                        normalization.method = "CLR",
                        margin = 2)

bcells[["percent_mt"]] <- PercentageFeatureSet(bcells, features = mt_genes)
bcells[["percent_ribo"]] <- PercentageFeatureSet(bcells, features = ribo_genes)
```

## QC


```{r, echo=FALSE, fig.height=3, fig.width=5, include=FALSE, eval=FALSE}
bcells_meta <- bcells@meta.data %>%
    rownames_to_column("barcode") %>%
    select(barcode, n_reads = nCount_RNA, 
           n_genes = nFeature_RNA, percent_mt, percent_ribo) %>%
    pivot_longer(-barcode, names_to = "feature")

bcells_meta %>%
    filter(feature %in% c("n_genes", "percent_mt")) %>%
    pivot_wider(names_from = feature, values_from = value) %>%
    ggplot(aes(percent_mt, n_genes)) +
    geom_jitter(size = .25, alpha = .25) +
    stat_density_2d(aes(fill = after_stat(level)), 
                    color = "slateblue4", alpha = .5,
                    geom = "polygon", contour = TRUE, 
                    bins = 10, n = 100) +
    geom_vline(xintercept = 10, linetype = 2, size = 1.25, color = "cornflowerblue") +
    geom_hline(yintercept = 500, linetype = 2, size = 1.25, color = "cornflowerblue") +
    scale_x_continuous(breaks = c(0, 5, 10, 20, 40, 60, 80)) +
    scale_y_continuous(breaks = seq(0, 1e4, by = 1e3)) +
    theme_classic() +
    theme(panel.grid.minor = element_blank(),
          legend.position = "none",
          panel.grid = element_blank()) +
    labs(x = "% Mitochondrial reads", y = "# of genes")

#bcells <- subset(bcells, subset = nFeature_RNA > 500 & percent_mt < 10)
```


Here we use the miQC package to model the percentage of mitochondrial reads and number of genes, in order to identify and remove compromised cells.


```{r, fig.height=3, fig.width=5}
bcells_sce <- as.SingleCellExperiment(bcells)
bcells_sce <- addPerCellQC(bcells_sce, subsets = list(mito = mt_genes))
model <- mixtureModel(bcells_sce)

plotFiltering(bcells_sce, model)
```

## Remove compromised cells

```{r}
bcells_sce <- filterCells(bcells_sce, model)
cells_keep <- rownames(colData(bcells_sce))
bcells <- subset(bcells, cells = cells_keep)
```

## Demultiplex cells based on HTO

```{r}
bcells <- HTODemux(bcells, assay = "HTO", positive.quantile = 0.99)

table(bcells$HTO_classification.global)

Idents(bcells) <- "HTO_maxID"
```

```{r, echo=FALSE}
stims <- c("Res00", "Res24", "IgG24", "IgG72", "RSQ24", "RSQ72")

stim_colors <- c("Res00" = "grey75", 
                 "Res24" = "grey60", 
                 "IgG24" = "gold1", 
                 "IgG72" = "gold3", 
                 "RSQ24" = "mediumpurple1", 
                 "RSQ72" = "mediumpurple4")

hto_data <- bcells@assays$HTO@data
hto_class_df <- tibble(barcode = colnames(hto_data),
                       maxid = bcells@meta.data$HTO_maxID)

hto_plot_df <- hto_data %>%
    as.data.frame() %>%
    rownames_to_column("hto") %>%
    pivot_longer(-hto, names_to = "barcode") %>%
    left_join(hto_class_df, by = "barcode") %>%
    select(maxid, hto, value) %>%
    mutate(maxid = factor(maxid, levels = stims),
           hto = factor(hto, levels = stims))

ggplot(hto_plot_df, aes(value, maxid, fill = maxid)) +
    geom_density_ridges(show.legend = FALSE, size = .25) +
    facet_wrap(~hto, nrow = 2) +
    scale_fill_manual(values = stim_colors) +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "HTO normalized levels", y = NULL)
```

```{r, echo=FALSE, fig.width=6.5, fig.height=4}
hto_global_df <- 
  tibble(barcode = colnames(hto_data),
         maxid = bcells@meta.data$HTO_maxID,
         classif = bcells@meta.data$HTO_classification.global) %>%
  count(maxid, classif) %>%
  group_by(maxid) %>%
  mutate(p = round(n/sum(n) * 100, 1),
         p = paste0(p, "%"),
         classif = factor(classif, levels = c("Negative", "Doublet", "Singlet")),
         maxid = factor(maxid, levels = stims)) %>%
  ungroup()

ggplot(hto_global_df, aes(classif, n, fill = maxid)) +
  geom_col(show.legend = FALSE, color = "black", size = .5) +
  geom_text(aes(label = p), size = 3, vjust = 0, nudge_y = 10) +
  scale_fill_manual(values = stim_colors) +
  facet_wrap(~maxid, nrow = 1, strip.position = "bottom") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid.major.x = element_blank(),
        panel.spacing = unit(0, "lines"),
        panel.border = element_blank()) +
  labs(x = NULL, y = "Number of droplets")
```

### HTO levels in Negative droplets

```{r, echo=FALSE}
stim_class_df <- tibble(barcode = colnames(bcells@assays$HTO@data),
                        stim = bcells@meta.data$HTO_maxID,
                        classif = bcells@meta.data$HTO_classification.global)

hto_df <- bcells@assays$HTO@data %>%
  as.data.frame() %>%
  rownames_to_column("hto") %>%
  as_tibble() %>%
  pivot_longer(-hto, names_to = "barcode") %>%
  left_join(stim_class_df)

hto_top <- hto_df %>%
  group_by(barcode, classif) %>%
  slice_max(n = 1, order_by = value) %>%
  select(barcode, classif, top_hto = hto)

hto_plot_df <- hto_df %>%
  left_join(hto_top, by = c("barcode", "classif")) %>%
  mutate(top_hto = factor(top_hto, levels = stims),
         stim = factor(stim, stims))

hto_plot_df %>%
  filter(classif == "Negative") %>%
  ggplot(aes(value)) +
  geom_density(aes(fill = stim), size = .25) +
  scale_fill_manual(values = stim_colors) +
  facet_wrap(~stim) +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  labs(x = "HTO normalized levels")
```

### HTO relative levels in Singlet, Doublet, and Negative droplets

```{r, echo=FALSE, fig.width=10, fig.height=8}
admix_singlet <- hto_plot_df %>%
  filter(classif == "Singlet") %>%
  ggplot(aes(reorder_within(barcode, by = value, within = stim), value)) +
  geom_col(aes(fill = hto), position = "fill", width = 1.01, show.legend = FALSE) +
  scale_fill_manual(values = stim_colors) +
  facet_grid(~stim, scales = "free_x", space = "free") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        strip.text = element_blank()) +
  labs(x = NULL, y = "HTO proportion", fill = "Stim", title = "Singlets")

admix_doublet <- hto_plot_df %>%
  filter(classif == "Doublet") %>%
  ggplot(aes(reorder_within(barcode, by = value, within = stim), value)) +
  geom_col(aes(fill = hto), position = "fill", width = 1.01, show.legend = FALSE) +
  scale_fill_manual(values = stim_colors) +
  facet_grid(~stim, scales = "free_x", space = "free") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        strip.text = element_blank()) +
  labs(x = NULL, y = "HTO proportion", fill = "Stim", title = "Doublets")

admix_negative <- hto_plot_df %>%
  filter(classif == "Negative") %>%
  ggplot(aes(reorder_within(barcode, by = value, within = stim), value)) +
  geom_col(aes(fill = hto), position = "fill", width = 1.01, show.legend = FALSE) +
  scale_fill_manual(values = stim_colors) +
  facet_grid(~stim, scales = "free_x", space = "free") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        strip.text = element_blank()) +
  labs(x = NULL, y = "HTO proportion", fill = "Stim", title = "Negatives")

plot_grid(admix_singlet, admix_doublet, admix_negative, ncol = 1)
```


## Extract Singlets

```{r}
Idents(bcells) <- "HTO_classification.global"

bcells_singlet <- subset(bcells, idents = "Singlet")

table(bcells_singlet@meta.data$HTO_maxID)[stims]
```
```{r, include=FALSE, eval=FALSE, echo=FALSE}
saveRDS(bcells_singlet, "./scDRS/data/expression/bcells_singlet_seurat.rds")
#bcells_singlet <- readRDS("./scDRS/data/expression/bcells_singlet_seurat.rds")
```




## Feature quantifications

```{r, echo=FALSE, fig.width=10, fig.height=3}
bcells_singlet_meta <- bcells_singlet@meta.data %>%
    rownames_to_column("barcode") %>%
    mutate(cell_class = bcells_singlet@meta.data$HTO_maxID,
           cell_class = factor(cell_class, levels = stims)) %>%
    select(barcode, cell_class, 
           n_reads = nCount_RNA, n_genes = nFeature_RNA, percent_mt, percent_ribo) %>%
    pivot_longer(-(barcode:cell_class), names_to = "feature")

ggplot(bcells_singlet_meta, aes(cell_class, value, fill = cell_class)) +
    geom_violin(size = .25) +
    scale_y_continuous(breaks = scales::pretty_breaks(7)) +
    scale_fill_manual(values = stim_colors) +
    facet_wrap(~feature, nrow = 1, scales = "free") +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.title = element_blank(),
          axis.ticks.x = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid.minor = element_blank()) +
    labs(fill = NULL)
```


## PCA

```{r}
bcells_singlet <- 
    FindVariableFeatures(bcells_singlet, 
                         nfeatures = 2000,
                         selection.method = "vst")

all_genes <- rownames(bcells_singlet)

bcells_singlet <- ScaleData(bcells_singlet, features = all_genes)

bcells_singlet <- RunPCA(bcells_singlet, features = VariableFeatures(bcells_singlet))
```


```{r, echo=FALSE, fig.height=7}
pca_cell_emb <- bcells_singlet@reductions$pca@cell.embeddings %>%
  as.data.frame() %>%
  rownames_to_column("barcode") %>%
  as_tibble() %>%
  left_join(stim_class_df) %>%
  select(barcode, stim, PC_1:PC_4)

pca_plot <- ggplot(pca_cell_emb, aes(PC_1, PC_2)) +
  geom_point(aes(color = stim), size = .5) +
  scale_color_manual(values = stim_colors) +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  guides(color = guide_legend(override.aes = list(size = 2)))

sdev_plot <- tibble(sdev = bcells_singlet@reductions$pca@stdev) %>%
    rownames_to_column("pc") %>%
    mutate(pc = fct_inorder(pc)) %>%
    filter(pc %in% 1:25) %>%
    ggplot(aes(pc, sdev)) +
    geom_line(aes(group = 1)) +
    theme_bw() +
    theme(panel.grid.minor = element_blank()) +
    labs(x = "Principal component", y = "Standard deviation")

pca_df <- bcells_singlet@reductions$pca@feature.loadings %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  as_tibble() 

loadings_plot <- pca_df %>%
  select(gene_id, PC_1, PC_2) %>%
  pivot_longer(-gene_id, names_to = "pc") %>%
  mutate(direction = case_when(sign(value) == 1 ~ "+",
                               sign(value) == -1 ~ "-",
                               sign(value) == 0 ~ "0")) %>%
  group_by(pc, direction) %>%
  top_n(15, abs(value)) %>%
  ungroup() %>%
  left_join(select(features_df, gene_id, gene_name)) %>%
  ggplot(aes(value, reorder_within(gene_name, value, pc))) +
  geom_col(aes(fill = direction), show.legend = FALSE) +
  scale_y_discrete(labels = function(x) sub("^([^_]+).*$", "\\1", x)) +
  scale_fill_manual(values = c("+" = "cornflowerblue", "-" = "tomato3")) +
  facet_wrap(~pc, scales = "free") +
  labs(x = "Loading", y = NULL) +
  theme_bw()

plot_grid(plot_grid(pca_plot, sdev_plot, rel_widths = c(1, .7)),
          loadings_plot,
          ncol = 1, rel_heights = c(.7, 1))
```


## Clustering

```{r}
# Find neighboring cells
bcells_singlet <- FindNeighbors(bcells_singlet, dims = 1:20)

# Cluster
bcells_singlet <- FindClusters(bcells_singlet, resolution = 0.25)
bcells_singlet <- FindClusters(bcells_singlet, resolution = 0.5)
bcells_singlet <- FindClusters(bcells_singlet, resolution = 0.8)
bcells_singlet <- FindClusters(bcells_singlet, resolution = 1.25)
```

## UMAP

```{r}
bcells_singlet <- RunUMAP(bcells_singlet, dims = 1:20)
```

### HTO classification

```{r, echo = FALSE, fig.height=4}
umap_df <- as.data.frame(bcells_singlet@reductions$umap@cell.embeddings) %>%
    as_tibble() %>%
    mutate(barcode = colnames(bcells_singlet@assays$RNA@data),
           cluster_0.25 = bcells_singlet@meta.data$RNA_snn_res.0.25,
           cluster_0.5 = bcells_singlet@meta.data$RNA_snn_res.0.5,
           cluster_0.8 = bcells_singlet@meta.data$RNA_snn_res.0.8,
           cluster_1.25 = bcells_singlet@meta.data$RNA_snn_res.1.25,
           stim = bcells_singlet@meta.data$HTO_maxID,
           stim = factor(stim, levels = stims),
           percent_mt = bcells_singlet@meta.data$percent_mt,
           percent_ribo = bcells_singlet@meta.data$percent_ribo,
           n_genes = bcells_singlet@meta.data$nFeature_RNA)

ggplot(umap_df, aes(UMAP_1, UMAP_2, color = stim)) +
    geom_point(size = 1) +
    scale_color_manual(values = stim_colors) +
    theme_bw() +
    theme(panel.grid = element_blank(),
          panel.border = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2)))
```

### Clustering at different resolutions

```{r, echo=FALSE, fig.height=5.5, fig.width=8}
cluster_df <- umap_df %>%
  pivot_longer(starts_with("cluster"), 
               names_to = "res", values_to = "cluster") %>%
  select(barcode, UMAP_1, UMAP_2, res, cluster) %>%
  mutate(res = sub("cluster_", "res = ", res))
  
cluster_labs <- cluster_df %>%
  group_by(res, cluster) %>%
  summarise_at(vars(UMAP_1, UMAP_2), median) %>%
  ungroup()


clustering_plot_list <- cluster_df %>%
  split(.$res) %>% 
  map(. %>% {
    data_df <- .
    labs_df <- filter(cluster_labs, res %in% unique(data_df$res))
    
    ggplot(data = data_df) +
      geom_point(aes(UMAP_1, UMAP_2, color = cluster), size = 1) +
      geom_text(data = labs_df, aes(UMAP_1, UMAP_2, label = cluster),
                color = "tomato", fontface = "bold") +
      scale_color_viridis_d() +
      facet_wrap(~res, nrow = 2) +
      theme_minimal() +
      theme(panel.grid = element_blank(),
            axis.title = element_blank(),
            axis.text = element_blank(),
            strip.text = element_text(size = 14),
            legend.key.size = unit(.1, "cm")) +
      guides(color = guide_legend(override.aes = list(size = 2)))
  })
   

plot_grid(plotlist = clustering_plot_list, ncol = 2)
```



## Marker genes for Seurat clusters (whole data, res = 1.25)


```{r}
cluster_markers <- 
    FindAllMarkers(bcells_singlet, 
                   only.pos = TRUE,
                   min.pct = 1/3,
                   logfc.threshold = 1) %>%
    as_tibble() %>%
    filter(p_val_adj < 0.01)
```  

## Top 5 marker genes per cluster

```{r, echo=FALSE, fig.height=12, fig.width=8}
top_cluster_markers <- cluster_markers %>%
  as_tibble() %>%
  left_join(select(features_df, 1:2), by = c("gene" = "gene_id")) %>%
  group_by(cluster) %>%
  top_n(5, avg_log2FC) %>%
  ungroup() %>%
  select(cluster, gene_name, gene_id = gene, avg_log2FC)

cluster_cell <- tibble(cluster_cell = Idents(bcells_singlet),
                       barcode = names(Idents(bcells_singlet)))
  
cluster_cell_gene_expr <- bcells_singlet@assays$RNA@data %>%
  .[unique(top_cluster_markers$gene_id), ] %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  as_tibble() %>%
  pivot_longer(-gene_id, names_to = "barcode", values_to = "logexpr")
  
top_cluster_marker_expr <- cluster_cell_gene_expr %>%
  left_join(cluster_cell, by = "barcode") %>%
  inner_join(top_cluster_markers, by = "gene_id")

top_cluster_marker_summary <- top_cluster_marker_expr %>%
  group_by(cluster = cluster_cell, gene_name) %>%
  summarise(prop_expr = mean(logexpr > 0),
            scale_expr = mean(logexpr)) %>%
  ungroup() %>%
  left_join(select(top_cluster_markers, cluster_top = cluster, gene_name, avg_log2FC)) %>%
  mutate_at(vars(cluster, cluster_top), factor) %>%
  arrange(cluster_top, avg_log2FC) %>%
  mutate(gene_name = fct_inorder(gene_name)) 


ggplot(top_cluster_marker_summary, aes(cluster, gene_name)) +
  geom_point(aes(size = prop_expr, fill = scale_expr), 
             color = "black", shape = 21) +
  scale_size(range = c(0.1, 4), labels = scales::percent) +
  scale_fill_viridis_c(option = "magma") +
  facet_wrap(~cluster_top, scales = "free", ncol = 2) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.text.x = element_text(size = 10),
        panel.grid = element_line(color = "grey96"),
        panel.border = element_blank(),
        legend.position = "top") +
  labs(x = NULL, y = NULL, 
       fill = "Scaled\nExpression", 
       size = "% of cells") +
   guides(fill = guide_colorbar(barheight = .5))
```



### Cell cycling

```{r, echo=FALSE, fig.height=5, fig.width=8}
mki_gene_df <- features_df %>%
    filter(gene_name == "MKI67") %>%
    select(gene_id, gene_name)

mki_gene_quant <- bcells_singlet@assays$RNA@data %>% 
    .[mki_gene_df$gene_id, ,drop = FALSE] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(mki_gene_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp")

ki67 <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(mki_gene_quant, by = "barcode") %>%
    ggplot(aes(UMAP_1, UMAP_2, color = gene_exp)) +
    geom_point(size = .75) +
    scale_color_viridis_c() +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "MKI67 RNA expression", color = NULL)

ngenes <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = n_genes)) +
    geom_point(size = .75) +
    scale_color_viridis_c(breaks = scales::pretty_breaks(10)) +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "Number of genes", color = NULL)

mt <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = percent_mt)) +
    geom_point(size = .75) +
    scale_color_viridis_c() +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "% Mitochondrial genes", color = NULL)

ribo <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = percent_ribo)) +
    geom_point(size = .75) +
    scale_color_viridis_c() +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "% Ribosomal genes", color = NULL)


plot_grid(ki67, ngenes, mt, ribo, nrow = 2)
```


### Downsampling

We see that RSQ 72hr is picked as a separate cluster with resolution = 0.6. In the full dataset, that does not happen even if we increase the resolution. Therefore, RSQ 72hr is not a separate cluster in the full dataset potentially due to low number of cells.


```{r}
set.seed(1)
sampled_cells <- 
  tibble(cell_barcode = colnames(bcells_singlet@assays$RNA@counts),
         cell_stim = bcells_singlet@meta.data$HTO_maxID) %>%
  select(cell_stim, cell_barcode) %>%
  group_by(cell_stim) %>%
  sample_n(178) %>%
  ungroup() %>%
  pull(cell_barcode)

bcells_singlet_downsamp <- subset(bcells_singlet, cells = sampled_cells)

table(bcells_singlet_downsamp@meta.data$HTO_maxID)[stims]

bcells_singlet_downsamp <- 
    FindVariableFeatures(bcells_singlet_downsamp, 
                         nfeatures = 1000,
                         selection.method = "vst")

bcells_singlet_downsamp <- 
    ScaleData(bcells_singlet_downsamp, features = VariableFeatures(bcells_singlet_downsamp))

bcells_singlet_downsamp <-
    RunPCA(bcells_singlet_downsamp, features = VariableFeatures(bcells_singlet_downsamp))
```

```{r}
# Find neighboring cells
bcells_singlet_downsamp <- FindNeighbors(bcells_singlet_downsamp, dims = 1:20)

# Cluster
bcells_singlet_downsamp <- FindClusters(bcells_singlet_downsamp, resolution = 0.6)

# UMAP
bbcells_singlet_downsamp <- RunUMAP(bcells_singlet_downsamp, dims = 1:20)
```

```{r, echo=FALSE, fig.width=10, fig.height=3}
umap_ds_df <- as.data.frame(bcells_singlet_downsamp@reductions$umap@cell.embeddings) %>%
    as_tibble() %>%
    mutate(barcode = colnames(bcells_singlet_downsamp@assays$RNA@data),
           cluster = bcells_singlet_downsamp@meta.data$seurat_clusters,
           stim = bcells_singlet_downsamp@meta.data$HTO_maxID,
           stim = factor(stim, levels = stims))

cell_class_ds_p <- ggplot(umap_ds_df, aes(UMAP_1, UMAP_2, color = stim)) +
    geom_point(size = 1) +
    scale_color_manual(values = stim_colors) +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "Hashtag classification")

seurat_cluster_ds_p <- ggplot(umap_ds_df, aes(UMAP_1, UMAP_2, color = cluster)) +
    geom_point(size = 1) +
    scale_color_viridis_d() +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "Seurat cluster res = 0.6")

plot_grid(cell_class_ds_p, seurat_cluster_ds_p, nrow = 1)
```






## B cell genes (RNA)

```{r, echo=FALSE, fig.width=9, fig.height=18}
bcell_genes <- 
    c("IGHD", "IGHM", "IGHE", "IGHG1", "IGHG2", "IGHG3", "IGHG4", 
      "CD19", "MS4A1", "CR2", "FCER2", "CD27", "CD38", "CD40", "CD69", "CD86",
      "ITGAX", "TBX21", "CXCR5", "HLA-DRA", 
      "TCL1A", "IL4R",
      "TNFSF13B", "TNFRSF13B", "TNFRSF17", "JCHAIN", "IL6")

bcell_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           gene_name %in% bcell_genes) %>%
    select(gene_id, gene_name)

bcell_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[bcell_genes_df$gene_id, ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(bcell_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp") %>%
    mutate(gene_name = factor(gene_name, levels = bcell_genes))

bcell_genes_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(bcell_genes_quant, by = "barcode") %>%
    split(.$gene_name) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = gene_exp)) +
            geom_point(size = .2) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5)) +
            facet_wrap(~gene_name) +
            theme_bw() +
            theme(panel.grid = element_blank(),
                  panel.border = element_blank(),
                  strip.background = element_blank(),
                  axis.line = element_blank(),
                  axis.ticks = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = bcell_genes_plot_list, ncol = 3)
```


## B cell genes (Protein)

```{r, echo=FALSE, fig.height=12, fig.width=9}
bcell_prots <- 
    c("IGHD" = "IgD", "IGHM" = "IgM",
      "CD19" = "CD19", "MS4A1" = "CD20", 
      "CR2" = "CD21", "FCER2" = "CD23",
      "CD27" = "CD27",
      "CD38" = "CD38", "CD40" = "CD40", 
      "CD69" = "CD69", "CD86" = "CD86",
      "ITGAX" = "CD11c", "CXCR5" = "CD185.or.CXCR5",
      "HLA-DRA" = "HLA-DR", 
      "TNFRSF13B" = "CD267.or.TACI",
      "TNFRSF13C" = "CD268.or.BAFF-R")

adt_quants <- bcells_singlet@assays$ADT@data[bcell_prots, ] %>%
    as.data.frame() %>%
    rownames_to_column("ab") %>%
    as_tibble() %>%
    pivot_longer(-ab, names_to = "barcode", values_to = "ab_level") %>%
    mutate(ab = factor(ab, levels = bcell_prots))

bcell_prots_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(adt_quants, by = "barcode") %>%
    split(.$ab) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = ab_level)) +
            geom_point(size = .2) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5,
                                                         barheight = 3)) +
            facet_wrap(~ab) +
            theme_bw() +
            theme(panel.grid = element_blank(),
                  panel.border = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank(),
                  strip.background = element_blank(),
                  axis.line = element_blank(),
                  axis.ticks = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = bcell_prots_plot_list, ncol = 3)

```



### IgD vs CD27


```{r, echo=FALSE, fig.height=2.5, fig.width=6}
igd_cd27_rna <- bcell_genes_quant %>%
  filter(gene_name %in% c("IGHD", "CD27")) %>%
  mutate(gene_name = recode(gene_name, "IGHD" = "IgD")) %>%
  select(barcode, gene_name, scale_exp = gene_exp) %>%
  left_join(select(umap_df, barcode, stim))

igd_cd27_prot <- adt_quants %>%
  filter(ab %in% c("IgD", "CD27")) %>%
  select(barcode, gene_name = ab, scale_exp = ab_level) %>%
  left_join(select(umap_df, barcode, stim))

bind_rows(mRNA = igd_cd27_rna, Protein = igd_cd27_prot, .id = "molecule") %>%
  pivot_wider(names_from = gene_name, values_from = scale_exp) %>%
  ggplot(aes(IgD, CD27)) +
  geom_point(size = .75, alpha = .5) +
  scale_color_manual(values = stim_colors) +
  facet_grid(molecule~stim) +
  theme_bw() +
  theme(panel.grid = element_blank())
```


## DN2 genes (Jenks et al. (2018); Fig 4-C)

```{r, echo=FALSE, fig.height=15, fig.width=9}
dn2_genes <- c("IL10RB", "ITGAX", "TBX21", "ZEB2", "SLAMF7", "TBKBP1",
               "IFNLR1", "SOX5", "DAPP1", "FCGR2B", "FCRL5", "FCRL3",
               "CD22", "IRF4", "IL10RA", "CD72", "BCL6", "IL21R", 
               "DUSP6", "KLF4")

dn2_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           gene_name %in% dn2_genes) %>%
    select(gene_id, gene_name)

dn2_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[dn2_genes_df$gene_id, , drop = FALSE] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(dn2_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp")

dn2_genes_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(dn2_genes_quant, by = "barcode") %>%
    split(.$gene_name) %>%
    .[dn2_genes] %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = gene_exp)) +
            geom_point(size = .2) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5)) +
            facet_wrap(~gene_name) +
            theme_minimal() +
            theme(panel.grid = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = dn2_genes_plot_list, ncol = 3)
```


## Lupus genes

```{r, echo=FALSE, fig.height=24, fig.width=10}
sle_genes <- 
    "https://www.nature.com/articles/ng.3434/tables/2" %>%
    read_html() %>%
    html_node("table") %>%
    html_table(header = TRUE, fill = TRUE) %>%
    select(gene = `Likely causal genec`) %>%
    slice(-1) %>%
    separate_rows(gene, sep = ",") %>%
    mutate(gene = trimws(gene)) %>%
    filter(gene != "") %>%
    pull(gene)

#sle_genes %>% write_lines("./scDRS/data/bentham_genes.txt")

sle_genes <- c(sle_genes, "ORMDL3")

sle_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           gene_name %in% sle_genes) %>%
    select(gene_id, gene_name)

sle_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[sle_genes_df$gene_id, ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(sle_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp") %>%
    group_by(gene_id) %>%
    filter(mean(gene_exp > 0) > .01) %>%
    ungroup()

sle_genes_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(sle_genes_quant, by = "barcode") %>%
    split(.$gene_name) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = gene_exp)) +
            geom_point(size = .1) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5)) +
            facet_wrap(~gene_name) +
            theme_minimal() +
            theme(panel.grid = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = sle_genes_plot_list, ncol = 4)

```



## TLR genes

```{r, echo=FALSE}
tlr_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("TLR\\d+$", gene_name) | gene_name == "MYD88") %>%
    select(gene_id, gene_name)

tlr_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[tlr_genes_df$gene_id, ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(tlr_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp") %>%
    mutate(gene_name = factor(gene_name, levels = c(paste0("TLR", 1:10), "MYD88")))

umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(tlr_genes_quant, by = "barcode") %>%
    ggplot(aes(UMAP_1, UMAP_2, color = gene_exp)) +
    geom_point(size = .1) +
    scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                          guide = guide_colorbar(barwidth = .5)) +
    facet_wrap(~gene_name) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          axis.title = element_blank(),
          axis.text = element_blank()) +
    labs(color = NULL)
```




## Find marker genes for each stim condition


```{r}
Idents(bcells_singlet) <- "HTO_maxID"

bcells_markers <- 
    FindAllMarkers(bcells_singlet, 
                   only.pos = TRUE,
                   min.pct = 1/3,
                   logfc.threshold = 1) %>%
    as_tibble()
```  



## Top 15 marker genes per cluster

```{r, echo=FALSE, fig.height=8, fig.width=8}
top_markers <- bcells_markers %>%
  as_tibble() %>%
  left_join(select(features_df, 1:2), by = c("gene" = "gene_id")) %>%
  group_by(cluster) %>%
  top_n(15, avg_log2FC) %>%
  ungroup() %>%
  select(cluster, gene_name, gene_id = gene, avg_log2FC)

cluster_cell <- tibble(cluster_cell = Idents(bcells_singlet),
                       barcode = names(Idents(bcells_singlet)))
  
cell_gene_expr <- bcells_singlet@assays$RNA@data %>%
  .[unique(top_markers$gene_id), ] %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  as_tibble() %>%
  pivot_longer(-gene_id, names_to = "barcode", values_to = "logexpr")
  
top_marker_expr <- cell_gene_expr %>%
  left_join(cluster_cell, by = "barcode") %>%
  inner_join(top_markers, by = "gene_id")

top_marker_summary <- top_marker_expr %>%
  group_by(cluster = cluster_cell, gene_name) %>%
  summarise(prop_expr = mean(logexpr > 0),
            scale_expr = mean(logexpr)) %>%
  ungroup() %>%
  left_join(select(top_markers, cluster_top = cluster, gene_name, avg_log2FC)) %>%
  mutate_at(vars(cluster, cluster_top), ~factor(., levels = stims)) %>%
  arrange(cluster_top, avg_log2FC) %>%
  mutate(gene_name = fct_inorder(gene_name)) 


ggplot(top_marker_summary, aes(cluster, gene_name)) +
  geom_point(aes(size = prop_expr, fill = scale_expr), 
             color = "black", shape = 21) +
  scale_size(range = c(0.1, 4), labels = scales::percent) +
  scale_fill_viridis_c(option = "magma") +
  facet_wrap(~cluster_top, scales = "free", ncol = 2) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.text.x = element_text(size = 10, angle = 45, hjust = 1),
        panel.grid = element_line(color = "grey96"),
        panel.border = element_blank()) +
  labs(x = NULL, y = NULL, 
       fill = "Scaled\nExpression", 
       size = "% of cells") +
   guides(fill = guide_colorbar(barwidth = .5))
```

## Marker genes IgG vs RSQ

### 24 hours

```{r}
bcells_markers_24 <- 
    FindMarkers(bcells_singlet, 
                   ident.1 = "IgG24",
                   ident.2 = "RSQ24",
                   only.pos = FALSE,
                   min.pct = 1/3,
                   logfc.threshold = 0.5) %>%
    rownames_to_column("gene") %>%
    as_tibble() %>%
    select(gene, avg_log2FC, IgG24 = pct.1, RSQ24 = pct.2, p = p_val_adj) %>%
    filter(p < 0.01)
```

```{r, echo=FALSE}
markers_timep_df <- bcells_markers_24 %>%
  mutate(cluster_top = ifelse(avg_log2FC > 0, "IgG24", "RSQ24")) %>%
  left_join(select(features_df, 1:2), by = c("gene" = "gene_id")) %>%
  pivot_longer(IgG24:RSQ24, names_to = "cluster", values_to = "pct")

cell_gene_expr_timep <- bcells_singlet@assays$RNA@data %>%
  .[unique(markers_timep_df$gene), ] %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  as_tibble() %>%
  pivot_longer(-gene_id, names_to = "barcode", values_to = "logexpr")

timep_gene_avg_expr <- cell_gene_expr_timep %>%
  left_join(cluster_cell) %>%
  filter(cluster_cell %in% c("IgG24", "RSQ24")) %>%
  group_by(cluster = cluster_cell, gene_id) %>%
  summarise(scale_expr = mean(logexpr)) %>%
  ungroup() %>%
  left_join(select(features_df, gene_id, gene_name)) %>%
  left_join(markers_timep_df, by = c("gene_id" = "gene", "gene_name", "cluster")) %>%
  arrange(avg_log2FC) %>%
  mutate(gene_name = fct_inorder(gene_name))
```

#### GWAS genes

```{r, echo = FALSE, fig.width=4, fig.height=3}
reported_genes <- read_lines("./reported_genes.tsv")

de_24_df <- timep_gene_avg_expr %>%
  filter(gene_name %in% reported_genes) %>%
  filter(! grepl("HLA", gene_name)) %>% 
  mutate_at(vars(cluster, cluster_top),
            ~recode(., "IgG24" = "BCR", "RSQ24" = "TRL7"))


ggplot(de_24_df, aes(cluster, gene_name)) +
   geom_point(aes(size = pct, fill = scale_expr), 
             color = "black", shape = 21) +
  scale_size(range = c(0.25, 5), labels = scales::percent) +
  scale_fill_viridis_c(option = "magma") +
  theme_bw() +
  theme(panel.background = element_blank(),
        plot.margin = margin(t = .5, b = 1, unit = "cm")) +
  labs(x = NULL, y = NULL, fill = "Scaled\nexpression", size = "% of\ncells") +
  guides(fill = guide_colorbar(barwidth = .5, barheight = 3),
         size = guide_legend(keyheight = .5))
```



## MAGMA

Scores taken from the scDRS figshare.

```{r, echo=FALSE, fig.height=6}
magma <- "./scDRS/data/MAGMA_v108_GENE_10_ZSTAT_for_scDRS.txt" %>%
  read.table() %>%
  rownames_to_column("gene_name") %>%
  select(gene_name, z = PASS_Lupus) %>%
  drop_na() %>%
  mutate(abs_z = abs(z))

magma_plot_df <- magma %>%
  top_n(50, abs_z) %>%
  arrange(abs_z) %>%
  mutate(gene_name = fct_inorder(gene_name))

ggplot(magma_plot_df, aes(abs_z, gene_name)) +
  geom_point(size = 4, alpha = .5) +
  theme_minimal() +
  theme(panel.grid.minor = element_blank()) +
  labs(x = "Absolute z-score", y = NULL)
```



## scDRS

```{r, echo=FALSE, fig.width=5, fig.height=3}
scdrs_scores <- read_tsv("./scDRS/results/Bentham_data_scDRSpaper/SLE.score.gz") %>%
  rename("barcode" = 1)

umap_scdrs_df <- umap_df %>%
  select(1:3) %>%
  left_join(scdrs_scores, by = "barcode")

ggplot(umap_scdrs_df, aes(UMAP_1, UMAP_2, color = norm_score)) +
    geom_point(size = .5) +
    scale_color_gradient2() +
    theme(panel.grid = element_blank(),
          panel.background = element_rect(fill = "grey")) +
    guides(color = guide_legend(override.aes = list(size = 2)))

```





