---
title: "README"
output: github_document
---

```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, comment = "#")
```

# Packages

```{r}
library(tidyverse)
library(Seurat)

# Plotting
library(ggbeeswarm)
library(ggridges)
library(MetBrewer)
```

```{r}
cellranger_dir <- 
    file.path("/lab-share/IM-Gutierrez-e2/Public/scRNA/SN0231064/KW9100_Maria",
              "210726_10X_KW9100-2_bcl/cellranger-6.0.1/GRCh38/BRI-1283/outs",
              "filtered_feature_bc_matrix")

features_df <- file.path(cellranger_dir, "features.tsv.gz") %>%
    read_tsv(col_names = c("gene_id", "gene_name", "phenotype"))

mt_genes <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("^MT-", gene_name)) %>%
    pull(gene_id)

ribo_genes <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("^RPS\\d+|^RPL\\d+", gene_name))

data10x <- Read10X(cellranger_dir, gene.column = 1)
```

```{r}
gene_exp <- data10x[["Gene Expression"]]

antibody <- data10x[["Antibody Capture"]] %>%
    .[!grepl("^Hashtag", rownames(.)), ] 

rownames(antibody) <- rownames(antibody) %>%
    sub("_prot$", "", .) %>%
    gsub("_", ".", .)

hashtags <- data10x[["Antibody Capture"]] %>%
    .[grepl("^Hashtag", rownames(.)), ]

rownames(hashtags) <- 
    c("IgG72", "RSQ72", "IgG24", "RSQ24", "Res24", "Res00")

bcells <- CreateSeuratObject(counts = gene_exp, project = "bcells")
bcells <- NormalizeData(bcells, normalization.method = "LogNormalize")
bcells <- FindVariableFeatures(bcells, selection.method = "mean.var.plot")
bcells <- ScaleData(bcells, features = VariableFeatures(bcells))

bcells[["ADT"]] <- CreateAssayObject(counts = antibody)
bcells[["HTO"]] <- CreateAssayObject(counts = hashtags)

Assays(bcells)

bcells <- NormalizeData(bcells, assay = "HTO", 
                        normalization.method = "CLR")

bcells <- NormalizeData(bcells, assay = "ADT", 
                        normalization.method = "CLR",
                        margin = 2)
```

## Demultiplex cells based on HTO

```{r}
bcells <- HTODemux(bcells, assay = "HTO", positive.quantile = 0.99)

table(bcells$HTO_classification.global)
```

```{r}
Idents(bcells) <- "HTO_maxID"

stims <- c("Res00", "Res24", "IgG24", "IgG72", "RSQ24", "RSQ72")

hto_data <- bcells@assays$HTO@data
hto_class_df <- tibble(barcode = colnames(hto_data),
                       maxid = bcells@meta.data$HTO_maxID)

hto_df <- hto_data %>%
    as.data.frame() %>%
    rownames_to_column("hto") %>%
    pivot_longer(-hto, names_to = "barcode") %>%
    left_join(hto_class_df, by = "barcode") %>%
    select(maxid, hto, value) %>%
    mutate(maxid = factor(maxid, levels = stims),
           hto = factor(hto, levels = stims))

ggplot(hto_df, aes(value, maxid, fill = maxid)) +
    geom_density_ridges(show.legend = FALSE) +
    facet_wrap(~hto) +
    scale_fill_manual(values = met.brewer("Austria", 6)) +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "Expression level", y = NULL)
```


## Extract Singlets

```{r}
Idents(bcells) <- "HTO_classification.global"

bcells_singlet <- subset(bcells, idents = "Singlet")
```


```{r}
bcells_singlet[["percent_mt"]] <- 
    PercentageFeatureSet(bcells_singlet, features = mt_genes)

bcells_meta <- bcells_singlet@meta.data %>%
    rownames_to_column("barcode") %>%
    select(barcode, orig.ident, nCount_RNA, nFeature_RNA, percent_mt) %>%
    pivot_longer(-(barcode:orig.ident), names_to = "feature")

ggplot(bcells_meta, aes(orig.ident, value)) +
    geom_violin(fill = "midnightblue", alpha = .5) +
    scale_y_continuous(breaks = scales::pretty_breaks(7)) +
    facet_wrap(~feature, nrow = 1, scales = "free") +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.title = element_blank(),
          axis.ticks.x = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank())
```

```{r}
n_genes <- Matrix::colSums(bcells_singlet@assays$RNA@counts > 0)

bcells_mt_ngenes <- bcells_meta %>%
    filter(feature == "percent_mt") %>%
    pivot_wider(names_from = feature, values_from = value) %>%
    left_join(tibble(barcode = names(n_genes),
                     n_genes = n_genes), 
              by = "barcode")

ggplot(bcells_mt_ngenes, aes(percent_mt, n_genes)) +
    geom_jitter(size = .25, alpha = .25) +
    geom_density2d_filled(alpha = .5) +
    geom_density_2d(size = .25, color = "black") +
    geom_vline(xintercept = c(5, 10, 20), linetype = 2) +
    scale_x_continuous(breaks = c(0, 5, 10, 20, 40, 60, 80)) +
    scale_y_continuous(breaks = seq(0, 1e4, by = 1e3)) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          legend.position = "none") +
    labs(x = "% Mitochondrial reads", y = "# of genes")
```

```{r}
bcells_singlet <- bcells_singlet %>%
    subset(subset = nFeature_RNA > 500 & percent_mt < 20)
```


```{r}
bcells_singlet <- 
    FindVariableFeatures(bcells_singlet, selection.method = "mean.var.plot")

bcells_singlet <- 
    ScaleData(bcells_singlet, features = VariableFeatures(bcells_singlet))

bcells_singlet <-
    RunPCA(bcells_singlet, features = VariableFeatures(bcells_singlet))

tibble(sdev = bcells_singlet@reductions$pca@stdev) %>%
    rownames_to_column("pc") %>%
    mutate(pc = fct_inorder(pc),
           var = sdev^2,
           prop_var = var/sum(var)) %>%
    filter(pc %in% 1:25) %>%
    ggplot(aes(pc, prop_var)) +
    geom_line(aes(group = 1)) +
    scale_y_continuous(labels = function(x) scales::percent(x, accuracy = 1L),
                       breaks = seq(0, .3, .05)) +
    theme_bw() +
    theme(panel.grid.minor = element_blank()) +
    labs(x = "Principal component", y = "Proportion of variance")

```

```{r}
bcells_singlet <- FindNeighbors(bcells_singlet, dims = 1:15)
bcells_singlet <- FindClusters(bcells_singlet, resolution = .15)
bcells_singlet <- RunUMAP(bcells_singlet, dims = 1:15)

as.data.frame(bcells_singlet@reductions$umap@cell.embeddings) %>%
    as_tibble() %>%
    mutate(cluster = bcells_singlet@meta.data$seurat_clusters,
           stim = bcells_singlet@meta.data$HTO_classification,
           stim = factor(stim, levels = stims)) %>%
    ggplot(aes(UMAP_1, UMAP_2, color = stim)) +
    geom_point(alpha = .75, size = .75) +
    scale_color_manual(values = met.brewer("Austria", 6)) +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

```


```{r, eval=FALSE}
bcells_markers <- 
    FindAllMarkers(bcells_singlet, 
                   only.pos = TRUE,
                   min.pct = 1/3,
                   logfc.threshold = .5)
```
