---
title: "README"
output: github_document
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

## B-Cells: Bulk RNA-seq

### Input data:

- 1 individual, 4 timepoints:
  - Resting for 16 hours
  - IgG stim for 24 hours
  - IgG stim for 72 hours
  - RSQ stim for 24 hours
  - RSQ stim for 72 hours
  
Fastq files located on directory:
/lab-share/IM-Gutierrez-e2/Public/B_cells/bulkTCpilot_1/34.198.31.178/210618_MG8989_fastq

### Methods

- Expression levels were estimated with Salmon (script `./salmon_quant.slurm`)


### Results

#### Overview of expression levels

```{r biotypes, echo=FALSE}
include_graphics("./plots/transcript_biotypes.png")
```

#### PCA of expression levels

```{r pca, echo=FALSE}
include_graphics("./plots/pca_bcell_expression.png")
```

#### Fold change in comparison with "resting"

```{r fc, echo=FALSE}
include_graphics("./plots/fc.png")
```

#### Subset of the plot above:

```{r fcsub, echo=FALSE}
include_graphics("./plots/fc_subset.png")
```

### TO DO:

- Run kallisto + sleuth once we have more samples, and call significant genes



## MGB Biobank analysis

### Input data

- 4921 individuals;
- ~79M variants.

### Methods

The procedure below is carried out by the `./process_vcf.slurm` script:

- VCF processing:
  - remove variants with any missing genotypes;
  - select only biallelic SNPs with MAF >= 0.1;
  - remove variants with pvalue < 0.05/79M for HWE test.
  
- 1000 Genomes data:
  - low coverage and exome data realigned to GRCh38 (not NYGC version);
  - same processing as above, except that I did not remove variants failing HWE.

- Match MGB and 1000G VCFs:
  - remove A/T and C/G genotypes due to potential strand ambiguity;
  - remove duplicate entries (these can be multiallelic variants);
  - select variants with the same position and alleles in both datasets;
  - filter both datasets for the common set of variants;
  - merge VCFs and run LD pruning for r2 > 0.2.

### PCA





