---
title: "README"
output: github_document
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = FALSE, warning=FALSE, message=FALSE)
```


## 0. General info

All the scripts with the `.slurm` extension can be submitted to the cluster using the command `sbatch script_name.slurm`.

For the `.sh` scripts, I execute them with the command `./script_name.sh` in an interactive job.


## 1. B cells: Bulk RNAseq

### 1.1. Input data:

- 1 individual sequenced at:
  - Resting for 16 hours
  - IgG stim for 24 hours
  - IgG stim for 72 hours
  - RSQ stim for 24 hours
  - RSQ stim for 72 hours
  
Fastq files located on directory:
/lab-share/IM-Gutierrez-e2/Public/B_cells/bulkTCpilot_1/34.198.31.178/210618_MG8989_fastq

### 1.2. Methods

#### 1.2.1. Alignment index

I obtained the genome sequence and annotation data from [Gencode](https://www.gencodegenes.org/human/release_38.html):

Using `wget **gencode_link**`, I downloaded the comprehensive gene annotation GTF file ("PRI") and the Genome sequence ("PRI"), which includes chromosomes and scaffolds. The files are in the `./data` directory.

Unfortunately, Gencode does not include a corresponding fasta file with transcript sequences for the "PRI" annotation, so I used RSEM to slice the Reference genome given the annotations, this way producing transcript sequences. For that I used the `./bcell_bulk/rsem.slurm` script.

I used Salmon to estimate expression. The script to create an index is `./bcell_bulk/salmon_index.slurm`.


#### 1.2.2. Expression estimation

Expression levels were estimated with script `./bcell_bulk/salmon_quant.slurm`.

#### 1.2.3. PCA on expression data

I used QTLtools to compute principal components from the expression matrix, using the script `./bcell_bull/pca.slurm`.

#### 1.2.4. Parsing result files

I compiled results from expression quantification, PCA, and other downstream analyses in R, with the script `./bcell_bulk/compile_results.R`.


#### 1.2.5. Plots

All the plots below were created with the script `./plot.R`.


### 1.3. Results

#### 1.3.1. Overview of expression levels

In the plot below we see the proportion of total expression attributed to each type of transcript, in Counts Per Million (CPM) and in Transcripts Per Million (TPM).

For the TPM plot, we see a jump in quantifications for rRNAs. But we need to keep in mind that these are very short RNAs, and a small increase in read counts can lead to large increases in TPM values.

```{r biotypes, echo=FALSE}
include_graphics("./plots/total_expression.png")
```

#### 1.3.2. PCA

PCA shows a separation of the RSQ and IgG treatments (PC1), and of the 24h/72h conditions (PC2).

```{r pca, echo=FALSE}
include_graphics("./plots/pca_bcell_expression.png")
```

#### 1.3.3. Fold change in comparison with "resting"

Here we have a scatter plot of CPM values in each condition against the log2 Fold-change in respect to the resting state.

Extreme values correspond to genes that have no expression in either the resting or test condition.

```{r fc, echo=FALSE}
include_graphics("./plots/foldchange.png")
```

#### 1.3.4. Subset of the plot above:

```{r fcsub, echo=FALSE}
include_graphics("./plots/foldchange_subset.png")
```

#### 1.3.5. Summary of direction of regulation

```{r updown}
include_graphics("./plots/foldchange_summary.png")
```


#### 1.3.6. Correlations with resting state:

```{r corr, echo=FALSE}
include_graphics("./plots/scatter_resting_conditions.png")
```

#### 1.3.7. Comparison of FC for each stimulant

```{r stims}
include_graphics("./plots/fc_24vs72.png")
```

#### 1.3.8. Expression of candidate SLE genes


We looked at the expression of candidate genes suggested by [Bentham et al. 2015, Table 2](http://www.nature.com/articles/ng.3434).

```{r sle}
include_graphics("./plots/bentham.png")
```

```{r sle_fc}
include_graphics("./plots/bentham_fc.png")
```


### 1.4. TO DO:

- Select genes in LD with SLE variants and look at their DE status;
- Run Maria's script to model expression given condition;
- Run kallisto + sleuth once we have more samples, and call significant genes.


## 2. MGB Biobank analysis

### 2.1. Goals

- Among the individuals in MGB Biobank, select those who carry more European ancestry;
- Among those, select individuals who are more heterozygous at SLE loci;
- We will use the SLE loci reported by Langefeld et al. (2017);
- Recruit these individuals to donate samples;
- Perform transcriptomic (ASE) and other analyses in multiple timepoints and stims;
- Identify dysregulated loci.

### 2.2. Input data

- MGB: 

| Batch | N     | Variants (MM) | Source          |
|:-----:|:-----:|:-------------:|:---------------:|
| 0401  | 4,921 | ~79.1         | MEGA_TopMed     |
| 0402  | 5,336 | ~80.1         | MEGAEX_TopMed   |
| 0403  | 4,780 | ~79.8         | MEG_A1_A_TopMed |
| 0404  | 5,016 | ~80.9         | MEG_A1_B_TopMed |
| 0405  | 5,491 | ~81.9         | MEG_C_TopMed    |
| 0406  | 5,143 | ~80.0         | MEG_D_TopMed    |
| 0407  | 4,847 | ~76.4         | MEG_E_TopMed    |
| 0408  | 866   | ~40.0         | MEG_X1_TopMed   |

- 1000 Genomes data:
  - ~2,500 individuals low coverage data realigned to GRCh38 (not NYGC version).

### 2.3. Methods

The workflow for the analyses below is described in `./mgb_biobank/README.md`.

- Selection of individuals who are genetically females
  - Compute individual-level heterozygosity with vcftools;
  - After visual inspection of the homozygosity distribution, select individuals with <0.985 homozygosity.

- VCF processing:
  - Remove variants with any missing genotypes;
  - Select only biallelic SNPs with MAF >= 0.1;
  - Remove A/T and C/G genotypes due to potential strand ambiguity;
  - Remove duplicates (these can be multiallelic variants or multiple variants with same position);
  - Select variants with the same position and alleles in both datasets;
  - Filter both datasets for the common set of variants;
  - Merge VCFs and run LD pruning for r2 < 0.1;
  - Concatenate VCFs for each chromosome into a single VCF.

- PCA:
  - plink pca

- ADMIXTURE
  - Unsupervised analysis on 1000 Genomes data;
  - Project MGB individuals onto 1000G reference panel.

- SLE risk variants
  - We take all variants at FDR<5% from [Langefeld et al. (2017)](http://www.nature.com/articles/ncomms16021);
  - For each individual, we compute an overall heterozygosity score at SLE variants.


### 2.4. Results

#### Select females

After visual inspection of the plot below, we select individuals with homozygosity < 0.985, which we assume to be females. That includes ~54.6% of the individuals.

```{r sex}
include_graphics("./plots/chrX_het.png")
```



#### Ancestry information from genotype data

##### PCA

The PCA plot shows the MGB individuals in comparison with the 1000 Genomes data. We can see that individuals are distributed according to the 5 main continental groups. 

```{r pca_genos, echo=FALSE}
include_graphics("./plots/pca.png")
```

We can really see that on the plot above, but most individuals appear to have high European ancestry.


```{r}
include_graphics("./plots/pca_histogram.png")
```



##### Selecting individuals with high European ancestry

In order to select individuals with European ancestry given PC coordinates, we computed the mean and SD for PC1 –PC4 in 1000 Genomes Europeans, and computed a z-score for each MGB individuals, selecting all individuals falling between -3 and +3 SDs for PC 1 and 2, and ±2 for PC 3 and 4.

```{r pca_selection}
include_graphics("./plots/pca_eur.png")
```


#### SLE variants

We selected SLE risk variants from Langefeld et al. and Bentham et al.

From Langefeld et al., we took all variants at FDR = 5%. From Bentham, we selected the variants which are not in LD with any variant in Langefeld et al. ($$r^2 < 0.6$$).

```{r bentham_ld}
include_graphics("./plots/sle_ld.png")
```

Then, we computed a heterozygosity score that corresponds to the number of variants at which the individuals are heterozygotes. We also computed a weighted score that is simply the number of alleles times the log(OR), summed over all variants. Since the number of alleles for heterozygotes is equal to 1, that corresponds to simply summing the log(OR) over all SLE SNPs.

Since we are not interested in the direction of effect, we converted all ORs < 1 to their reciprocal (1/OR).

This is the relationship between the two scores:

```{r het_score}
include_graphics("./plots/het_scores_points.png")
```

This are the distributions according the European ancestry:

```{r het_score_dens}
include_graphics("./plots/het_score_density.png")
```

We can see that heterozygosity increases with European ancestry. We can also see that if we color the PCA plot by heterozygosity.

```{r pca_het_score}
include_graphics("./plots/pca_het_scores.png")
```
