---
title: "Splicing"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, dpi = 300)
```


```{r}
library(tidyverse)
library(tidytext)
library(ggsci)
library(ggrepel)
library(cowplot)
```



```{r}
read_leaf <- function(cell_type) {
    
    sig <- 
        file.path("./results", cell_type, "leafcutter_cluster_significance.txt") %>%
        read_tsv() %>%
        separate(cluster, c("chr", "cluster"), sep = ":") %>%
        select(cluster, p.adjust, genes)

    eff <- 
        file.path("./results", cell_type, "leafcutter_effect_sizes.txt") %>%
        read_tsv() %>%
        mutate(cluster = sub("^.*(clu.*)$", "\\1", intron)) %>%
        select(cluster, logef, deltapsi)

    inner_join(sig, eff)
}
```


```{r}
cell_types <- c("rN", "aN", "T3", "SM", "DN") %>%
    setNames(., .)

leaf_df <- map_df(cell_types, read_leaf, .id = "cell_type") %>%
    group_by(cell_type, cluster) %>%
    slice(which.max(abs(deltapsi))) %>%
    ungroup()

dtu <- 
    file.path("./results", cell_types, "dtu_perGene.tsv") %>%
    setNames(cell_types) %>%
    map_df(~read_tsv(.) %>%
               select(gene_name, adj_pvalue),
           .id = "cell_type")
```


## Differential splicing events between patients and controls in Scharer dataset


```{r, echo = FALSE, fig.height=8, fig.width=10}
labs_df <- leaf_df %>%
    filter(p.adjust < 0.05) %>%
    group_by(cell_type) %>%
    filter(abs(deltapsi) >= 0.4 |
               -log10(p.adjust) >= quantile(-log10(p.adjust), 0.99)) %>%
    ungroup() %>%
    mutate(gene_label = str_trunc(genes, 10),
           gene_label = replace_na(gene_label, "N/A"))

ggplot(leaf_df, aes(deltapsi, -log10(p.adjust))) +
    geom_point(aes(color = p.adjust < .05), size = .7, show.legend = FALSE) +
    geom_text_repel(data = labs_df, aes(label = gene_label),
                    size = 2, segment.size = .2) +
    scale_color_manual(values = c("FALSE" = "grey85", "TRUE" = "grey60")) +
    scale_x_continuous(limits = c(-.75, .75)) +
    facet_wrap(~cell_type, ncol = 2, scales = "free") +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "Delta PSI", 
         y = "-log10(p-value)",
         color = "Significant?")
```

## GO enrichment

GO enrichment analysis on genes with at least one cluster with absolute delta PSI > 0.1.


```{r}
go <- read_tsv("./results/enrichment.tsv") %>%
    arrange(Description)
```


```{r, fig.height=7, fig.width=8}
go_top15 <- go %>%
    group_by(cell_type) %>%
    top_n(15, -log10(qvalue)) %>%
    ungroup() %>%
    mutate(cell_type = factor(cell_type, levels = c("rN", "aN", "T3", "SM", "DN"))) %>%
    arrange(cell_type, qvalue)

goterms <- distinct(go_top15, Description) %>%
    pull(Description)

go_plot <- go_top15 %>%
    mutate(Description = factor(Description, levels = rev(goterms))) %>%
    ggplot(aes(cell_type, Description)) +
        geom_point(aes(size = Count, fill = -log10(qvalue)), shape = 21) +
        scale_fill_viridis_c(option = "magma") +
        theme_minimal() +
        theme(axis.title = element_blank(),
              legend.position = "top") +
        labs(size = "# Genes:") +
        guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.5),
               size = guide_legend(direction = "horizontal"))

plot_grid(get_legend(go_plot), 
          go_plot + theme(legend.position = "none"),
          ncol = 1,
          rel_heights = c(0.1, 1))
             
```

## Comparison with differential transcript usage obtained with the DEXSeq pipeline


```{r, fig.height=10, fig.width=10}
leaf_uniq <- leaf_df %>%
    separate_rows(genes, sep = ",") %>%
    group_by(cell_type, genes) %>%
    slice(which.min(p.adjust)) %>%
    ungroup() %>%
    select(cell_type, gene_name = genes, adj_pvalue_leaf = p.adjust)

dtu_uniq <- dtu %>%
    group_by(cell_type, gene_name) %>%
    slice(which.min(adj_pvalue)) %>%
    ungroup() %>%
    select(cell_type, gene_name, adj_pvalue_dtu = adj_pvalue)

leaf_dtu <- full_join(leaf_uniq, dtu_uniq, by = c("cell_type", "gene_name")) %>%
    mutate_at(vars(adj_pvalue_leaf, adj_pvalue_dtu), ~replace_na(., 1)) %>%
    mutate(col = case_when(adj_pvalue_leaf < 0.05 & adj_pvalue_dtu < 0.05 ~ "TRUE",
                           TRUE ~ "FALSE"))

top_leaf_dtu <- leaf_dtu %>%
    group_by(cell_type) %>%
    filter(abs(-log10(adj_pvalue_leaf)) >= quantile(-log10(adj_pvalue_leaf), 0.995) |
               -log10(adj_pvalue_dtu) >= quantile(-log10(adj_pvalue_dtu), 0.995)) %>%
    ungroup() %>%
    mutate(gene_name = replace_na(gene_name, "N/A"))

ggplot(leaf_dtu, aes(-log10(adj_pvalue_leaf), -log10(adj_pvalue_dtu))) +
    geom_point(aes(color = col), alpha = .5, show.legend = FALSE) +
    geom_label_repel(data = top_leaf_dtu,
                    aes(label = gene_name),
                    force = .5,
                    size = 2,
                    label.size = NA,
                    label.padding = unit(0.1, "lines"),
                    alpha = .75,
                    segment.color = "grey35") +
    scale_color_manual(values = c("TRUE" = "black", "FALSE" = "grey")) +
    facet_wrap(~cell_type, ncol = 2, scales = "free") +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "-log10(p) [Leafcutter]", y = "-log10(p) [DTU]")
```

## Differentially spliced genes that are also GWAS genes


```{r, fig.width=6, fig.height=6}
gwas <- read_tsv("../bcell_scrna/reported_genes.tsv") %>%
    separate_rows(author, snp, p, sep = ",")
    
leaf_filtered <- read_tsv("./results/filtered_leafcutter.tsv")
sqtl <- read_tsv("./mu2021_data/filtered_sqtl.tsv", col_types = c(sid = "c"))

inner_join(leaf_filtered, gwas, by = c("genes" = "gene")) %>%
    arrange(desc(genes), author) %>%
    mutate_at(vars(genes, author), fct_inorder) %>%
    ggplot(aes(author, genes)) +
    geom_point(aes(color = deltapsi), size = 4) +
    scale_color_gradient2(guide = guide_colorbar(barwidth = .5)) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
          panel.grid = element_line(color = "grey96")) +
    labs(x = NULL, y = NULL)
```

```{r, fig.width=10, eval=FALSE}
inner_join(leaf_filtered, sqtl, by = "genes") %>%
    arrange(desc(genes), dataset) %>%
    mutate_at(vars(genes, dataset), fct_inorder) %>%
    ggplot(aes(genes, dataset)) +
    geom_point(aes(color = deltapsi), size = 4) +
    scale_color_gradient2(guide = guide_colorbar(barheight = .5)) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
          legend.position = "top",
          panel.grid = element_line(color = "grey96")) +
    labs(x = NULL, y = NULL)
```


