---
title: "Splicing"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, dpi = 300)
```


```{r}
library(tidyverse)
library(tidytext)
library(ggsci)
library(ggrepel)
library(cowplot)
```


```{r, fig.width=10, fig.height=3}
mapping_stats <- 
    c("/lab-share/IM-Gutierrez-e2/Public/External_datasets/Andreoletti/qc/multiqc_data/multiqc_fastqc.txt",
      "/lab-share/IM-Gutierrez-e2/Public/External_datasets/Barnas/qc/multiqc_data/multiqc_fastqc.txt",
      "/lab-share/IM-Gutierrez-e2/Public/External_datasets/Scharer/RNAseq/qc/multiqc_data/multiqc_fastqc.txt") %>%
    setNames(c("Andreoletti", "Barnas", "Scharer")) %>%
    map_df(~read_tsv(.) %>%
               select(Sample, "Total Sequences", total_deduplicated_percentage) %>%
               mutate(unique = `Total Sequences` * total_deduplicated_percentage / 100,
                      duplicate = `Total Sequences` - unique) %>%
               select(Sample, unique, duplicate), .id = "dataset") %>%
    pivot_longer(unique:duplicate, names_to = "type", values_to = "n")
    
mapping_stats %>%
    group_by(dataset, Sample) %>%
    mutate(total = sum(n)) %>%
    ungroup() %>%
    arrange(dataset, total) %>%
    mutate(Sample = fct_inorder(Sample)) %>%
    ggplot(aes(Sample, n, fill = type)) +
    geom_col(width = 1.05) +
    scale_y_continuous(labels = scales::comma) +
    scale_fill_manual(values = c("unique" = "lightskyblue", "duplicate" = "grey30")) +
    facet_grid(.~dataset, space = "free", scales = "free_x") +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          panel.grid = element_blank()) +
    labs(x = NULL, y = "Total reads", fill = NULL)



```


```{r, fig.height = 3, fig.width = 6}
get_stats <- function(dataset) {
    file.path("../read_mapping", dataset, "mapping") %>%
    list.files(pattern =  "*_Log.final.out", full.names = TRUE) %>%
    setNames(., str_remove(basename(.), "_Log.final.out")) %>%
    map_df(. %>%
	   read_lines() %>%
	   keep(~grepl("(Uniquely mapped reads %)|(% of reads mapped to multiple)|(% of reads unmapped)", .)) %>%
	   map_chr(trimws) %>%
	   str_split("\t") %>%
	   map_df(~tibble(stat = .[1], pct = .[2])) %>%
	   mutate(pct = parse_number(pct),
	          stat = case_when(grepl("^Uniquely", stat) ~ "uniq",
	                           grepl("multiple", stat) ~ "multi",
	                           grepl("unmapped", stat) ~ "unmap"),
	          stat = factor(stat, levels = c("uniq", "multi", "unmap"))) %>%
	   group_by(stat) %>%
	   summarise(pct = sum(pct)/100L) %>%
	   ungroup(), 
       .id = "sampleid")
}

stats_df <- bind_rows("Scharer" = get_stats("scharer"),
                      "Barnas" = get_stats("barnas"),
                      "Andreoletti" = get_stats("andreoletti"),
                      .id = "dataset")

ggplot(stats_df, aes(x = stat, y = pct)) +
    geom_violin(color = "black", fill = "black", size = .25) +
    scale_y_continuous(labels = scales::percent) +
    facet_wrap(~dataset) +
    theme_bw() +
    theme(panel.grid = element_line(color = "grey96")) +
    labs(x = NULL, y = NULL)
```


```{r}
cell_types_scharer <- c("rN", "aN", "T3", "SM", "DN") %>%
    setNames(., .)

leaf_scharer_df <- read_tsv("./results/scharer/leafcutter_filtered.tsv") %>%
    mutate(cell_type = factor(cell_type, levels = cell_types_scharer))

leaf_barnas_df <- read_tsv("./results/barnas/leafcutter_filtered.tsv")

leaf_and_df <- read_tsv("./results/andreoletti/leafcutter_filtered.tsv")

dtu_scharer <- 
    file.path("./results/scharer", cell_types_scharer, "dtu_perGene.tsv") %>%
    setNames(cell_types_scharer) %>%
    map_df(~read_tsv(.) %>%
               select(gene_name, adj_pvalue),
           .id = "cell_type")
```


## Differential splicing events between patients and controls in Scharer dataset

```{r, fig.height=8, fig.width=8}
labs_df <- leaf_scharer_df %>%
    filter(p.adjust < 0.05) %>%
    group_by(cell_type) %>%
    filter(abs(deltapsi) >= quantile(abs(deltapsi), 0.98) |
           -log10(p.adjust) >= quantile(-log10(p.adjust), 0.98)) %>%
    ungroup() %>%
    mutate(gene_label = str_trunc(genes, 10),
           gene_label = replace_na(gene_label, "N/A"))

ggplot(leaf_scharer_df, aes(deltapsi, -log10(p.adjust))) +
    geom_point(aes(color = p.adjust < .05), size = .7, show.legend = FALSE) +
    geom_text_repel(data = labs_df, aes(label = gene_label),
                    size = 2, segment.size = .2) +
    scale_color_manual(values = c("FALSE" = "grey85", "TRUE" = "grey60")) +
    scale_x_continuous(limits = c(-.75, .75)) +
    facet_wrap(~cell_type, ncol = 2, scales = "free") +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "Delta PSI", 
         y = "-log10(p-value)",
         color = "Significant?")
```


## Comparison between Scharer et al and Barnas et al 

```{r, fig.height=4, fig.width=8}
naive_dn <- 
    bind_rows(scharer = filter(leaf_scharer_df, cell_type %in% c("rN", "DN")),
	      barnas = filter(leaf_barnas_df, cell_type %in% c("Naive", "DN")),
	      .id = "dataset") %>%
    mutate(cell_type = recode(cell_type, "rN" = "Naive"),
	   abs_deltapsi = abs(deltapsi)) %>%
    select(dataset, cell_type, genes, abs_deltapsi) %>%
    separate_rows(genes, sep = ",") %>%
    group_by(dataset, cell_type, genes) %>%
    slice(which.max(abs_deltapsi)) %>%
    ungroup() %>%
    pivot_wider(names_from = dataset, values_from = abs_deltapsi) %>%
    mutate_at(vars(barnas:scharer), ~replace_na(., 0)) %>%
    filter(!is.na(genes))

labs_naivedn <- naive_dn %>%
    group_by(cell_type) %>%
    filter(barnas >= quantile(barnas, 0.995) |
           scharer >= quantile(scharer, 0.995) |
	   barnas * scharer >= quantile(barnas * scharer, 0.995)) %>%
    ungroup() %>%
    mutate(gene_label = str_trunc(genes, 10),
           gene_label = replace_na(gene_label, "N/A"))

ggplot(naive_dn, aes(x = scharer, y = barnas)) +
    geom_point(alpha = .5) +
    geom_text_repel(data = labs_naivedn, aes(label = gene_label),
                    size = 2, segment.size = .2) +
    facet_wrap(~cell_type) +
    theme_bw()
```



## Comparison with differential transcript usage obtained with the DEXSeq pipeline


```{r, fig.height=8, fig.width=8}
leaf_uniq <- leaf_scharer_df %>%
    separate_rows(genes, sep = ",") %>%
    group_by(cell_type, genes) %>%
    slice(which.min(p.adjust)) %>%
    ungroup() %>%
    select(cell_type, gene_name = genes, adj_pvalue_leaf = p.adjust)

dtu_uniq <- dtu_scharer %>%
    group_by(cell_type, gene_name) %>%
    slice(which.min(adj_pvalue)) %>%
    ungroup() %>%
    select(cell_type, gene_name, adj_pvalue_dtu = adj_pvalue)

leaf_dtu <- full_join(leaf_uniq, dtu_uniq, by = c("cell_type", "gene_name")) %>%
    mutate_at(vars(adj_pvalue_leaf, adj_pvalue_dtu), ~replace_na(., 1)) %>%
    mutate(col = case_when(adj_pvalue_leaf < 0.05 & adj_pvalue_dtu < 0.05 ~ "TRUE",
                           TRUE ~ "FALSE"))

top_leaf_dtu <- leaf_dtu %>%
    group_by(cell_type) %>%
    filter(abs(-log10(adj_pvalue_leaf)) >= quantile(-log10(adj_pvalue_leaf), 0.999) |
               -log10(adj_pvalue_dtu) >= quantile(-log10(adj_pvalue_dtu), 0.999)) %>%
    ungroup() %>%
    mutate(gene_name = replace_na(gene_name, "N/A"))

ggplot(leaf_dtu, aes(-log10(adj_pvalue_leaf), -log10(adj_pvalue_dtu))) +
    geom_point(aes(color = col), alpha = .5, show.legend = FALSE) +
    geom_label_repel(data = top_leaf_dtu,
                    aes(label = gene_name),
                    force = .5,
                    size = 2,
                    label.size = NA,
                    label.padding = unit(0.1, "lines"),
                    alpha = .75,
                    segment.color = "grey35") +
    scale_color_manual(values = c("TRUE" = "black", "FALSE" = "grey")) +
    facet_wrap(~cell_type, ncol = 2, scales = "free") +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "-log10(p) [Leafcutter]", y = "-log10(p) [DTU]")
```

## Exploring why AS sometimes does not lead to DTU


### AS in HMGB1, EIF4A2, and RPL15

```{r, fig.height=8}
annots <- read_rds("./plot_data/transcripts_annot.rds") %>%
    mutate(transcript_id = str_remove(transcript_id, "\\.\\d+$"))

load("./results/scharer/DN/DN.Rdata")

intron_df <- as_tibble(introns) %>%
    filter(gene %in% c("HMGB1", "EIF4A2", "RPL15")) %>%
    extract(clusterID, c("clusterID", "strand"), "(clu_.+)_([+-])") %>%
    filter(gene == "EIF4A2" & strand == "+" |
           gene == "HMGB1" & strand == "-" | 
           gene == "RPL15" & strand == "+") %>%
    group_by(clusterID) %>%
    mutate(i = max(abs(deltapsi))) %>%
    group_by(gene) %>%
    filter(i == max(i)) %>%
    ungroup() %>%
    select(-i)

plot_df <- annots %>% 
    filter(gene_name %in% c("HMGB1", "EIF4A2", "RPL15")) %>%
    group_by(transcript_id) %>%
    mutate(i = row_number(),
           col = case_when(feature == "intron" & i == min(i) ~ NA_character_,
                           feature == "intron" & i == max(i) ~ NA_character_,
                           feature == "exon" ~ "exon",
                           TRUE ~ "intron")) %>%
    ungroup()

as_ifoforms <-
    tribble(~gene_name, ~transcript_id,
            "HMGB1", "ENST00000399489",
            "EIF4A2", "ENST00000494445",
            "EIF4A2", "ENST00000425053")

intron_curves <- intron_df %>%
    select(gene_name = gene, start, end) %>%
    group_by(gene_name) %>%
    mutate(event = letters[1:n()]) %>%
    ungroup() %>%
    left_join(as_ifoforms)

  
plot_df %>%
    filter((gene_name == "HMGB1" & start > 30460000 & end < 30463000) |
           (gene_name == "EIF4A2" & start > 186787500 & end < 186790000) |
           (gene_name == "RPL15")) %>%
    filter(!is.na(col)) %>%
    ggplot() +
    geom_segment(aes(x = start, xend = end, 
                     y = transcript_id, yend = transcript_id,
                     color = as.factor(col)), 
                 size = 5) +
    geom_curve(data = intron_curves, 
               aes(x = start, xend = end, 
                   y = transcript_id, yend = transcript_id), 
               curvature = -.1) +
    scale_color_manual(values = c("intron" = "grey80", "exon" = "blue"),
                       na.translate = FALSE) +
    facet_wrap(~gene_name, scales = "free", ncol = 1) +
    theme_bw() +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank()) +
    labs(x = "Position", y = NULL, color = "Feature")
```


### Expression levels for the AS isoforms in HMGB1 and EIF4A2

```{r, fig.height=3, fig.width=10}
sample_annot <- 
    "../../../External_datasets/Scharer/RNAseq/metadata/file_description.tsv" %>%
    read_tsv() %>%
    select(sample_id = run, status, cell_subtype) %>%
    mutate(status = ifelse(grepl("SLE", status), "SLE", status))

transcript_gene <- 
    "../../../References/Annotations/hsapiens/gencode.v38.txToGene.tsv" %>%
    read_tsv()

scharer_tpm <- 
    "../../scharer/compiled_expression_humangenes.tsv" %>%
    read_tsv() %>%
    pivot_longer(-transcript_id, names_to = "sample_id", values_to = "tpm") %>%
    left_join(transcript_gene, by = "transcript_id") %>%
    left_join(sample_annot, by = "sample_id") %>%
    select(sample_id, cell_subtype, status, transcript_id, gene_id, gene_name, tpm) %>%
    mutate(transcript_id = str_remove(transcript_id, "\\.\\d+$"))
    
scharer_plot_df <- scharer_tpm %>%
    filter(grepl("DN2", cell_subtype),
           gene_name %in% c("HMGB1", "EIF4A2", "RPL15")) 
    
top_expressed <- scharer_plot_df %>%
    group_by(gene_name, transcript_id) %>%
    summarise(m = mean(tpm)) %>%
    ungroup() %>%
    group_by(gene_name) %>%
    top_n(8, m) %>%
    ungroup() %>%
    select(-m)

scharer_plot_df %>%
    inner_join(top_expressed) %>%
    ggplot(aes(y = reorder(transcript_id, tpm, "median"), x = tpm)) +
    geom_boxplot(aes(fill = status), outlier.color = NA, alpha = .5) +
    geom_point(aes(group = status), size = .5, position = position_dodge(width = 0.75)) +
    scale_fill_manual(values = c("Healthy" = "cornflowerblue", "SLE" = "tomato3")) +
    facet_wrap(~gene_name, scales = "free", ncol = 3) +
    theme_bw() +
    theme(panel.grid.major.y = element_blank()) +
    labs(x = "Transcripts per Million", y = NULL)
```



## GO enrichment

GO enrichment analysis on genes with at least one cluster with absolute delta PSI > 0.1.


```{r}
go <- read_tsv("./results/scharer/enrichment.tsv") %>%
    arrange(Description)
```


```{r, fig.height=7, fig.width=8}
go_top15 <- go %>%
    group_by(cell_type) %>%
    top_n(15, -log10(qvalue)) %>%
    ungroup() %>%
    mutate(cell_type = factor(cell_type, levels = c("rN", "aN", "T3", "SM", "DN"))) %>%
    arrange(cell_type, qvalue)

goterms <- distinct(go_top15, Description) %>%
    pull(Description)

go_plot <- go_top15 %>%
    mutate(Description = factor(Description, levels = rev(goterms))) %>%
    ggplot(aes(cell_type, Description)) +
        geom_point(aes(size = Count, fill = -log10(qvalue)), shape = 21) +
        scale_fill_viridis_c(option = "magma") +
        theme_minimal() +
        theme(axis.title = element_blank(),
              legend.position = "top") +
        labs(size = "# Genes:") +
        guides(fill = guide_colorbar(direction = "horizontal", barheight = 0.5),
               size = guide_legend(direction = "horizontal"))

plot_grid(get_legend(go_plot), 
          go_plot + theme(legend.position = "none"),
          ncol = 1,
          rel_heights = c(0.1, 1))
             
```


## Differentially spliced genes that are also GWAS genes

```{r, fig.width=6, fig.height=4}
gwas <- read_tsv("../bcell_scrna/reported_genes.tsv") %>%
    separate_rows(author, snp, p, sep = ",")
    
leaf_filtered <- read_tsv("./results/scharer/leafcutter_filtered_significant.tsv")
sqtl <- read_tsv("./mu2021_data/filtered_sqtl.tsv", col_types = c(sid = "c"))

inner_join(leaf_filtered, gwas, by = c("genes" = "gene")) %>%
    arrange(desc(genes), author) %>%
    mutate_at(vars(genes, author), fct_inorder) %>%
    ggplot(aes(author, genes)) +
    geom_point(aes(color = deltapsi), size = 3) +
    scale_color_gradient2(guide = guide_colorbar(barwidth = .5)) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
          axis.text.y = element_text(size = 7),
          panel.grid = element_line(color = "grey96")) +
    labs(x = NULL, y = NULL)
```

## Differentially spliced genes with sQTL-GWAS co-localizations in Mu et al. (2021)
```{r, fig.height=3}
sqtl_coloc <- read_tsv("./plot_data/coloc_sqtl.tsv") %>%
    unite("dataset", c("study", "cell"), sep = "-") %>%
    select(dataset, intron, gene)

coloc_df <- leaf_scharer_df %>%
    select(cluster, genes, p.adjust, deltapsi) %>%
    filter(p.adjust < 0.05, abs(deltapsi) > 0.1) %>%
    separate_rows(genes, sep = ",") %>%
    mutate(genes = trimws(genes)) %>%
    inner_join(sqtl_coloc, by = c("genes" = "gene"))

ggplot(coloc_df, aes(dataset, genes)) +
    geom_point(aes(color = deltapsi), size = 4) +
    scale_color_gradient2(guide = guide_colorbar(barwidth = .5)) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
          panel.grid = element_line(color = "grey96")) +
    labs(x = NULL, y = NULL)
```

## Andreoletti et al. 

```{r, fig.height=4, fig.width=5}
labs_and_df <- leaf_and_df %>%
    group_by(cell_type) %>%
    filter(abs(deltapsi) >= quantile(abs(deltapsi), 0.995) |
           p.adjust < 0.05) %>%
    ungroup() %>%
    mutate(gene_label = str_trunc(genes, 10),
           gene_label = replace_na(gene_label, "N/A"))

ggplot(leaf_and_df, aes(deltapsi, -log10(p.adjust))) +
    geom_point(aes(color = p.adjust < .05), size = 1, show.legend = FALSE) +
    geom_text_repel(data = labs_and_df, aes(label = gene_label),
                    size = 2, segment.size = .1) +
    scale_color_manual(values = c("FALSE" = "grey85", "TRUE" = "grey40")) +
    facet_wrap(~cell_type) +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "Delta PSI", 
         y = "-log10(p-value)",
         color = "Significant?")
```

