---
title: "CITE-seq Pilot"
output: github_document
---

```{r setup, include=FALSE}
library(unix)
rlimit_as(1e12)

library(knitr)
opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, comment = "#",
               dpi = 300)
options(width = 999)
```

## Packages

```{r}
# single-cell data analysis
library(Seurat)
library(miQC)
library(scater)

# Data wrangling
library(tidyverse)
library(rvest)

# Plotting
library(tidytext)
library(ggridges)
library(RColorBrewer)
library(cowplot)
```


## Cell Ranger data

```{r}
cellranger_dir <- 
    file.path("/lab-share/IM-Gutierrez-e2/Public/Lab_datasets/CITEseq_pilot", 
              "SN0231064/KW9100_Maria/210726_10X_KW9100-2_bcl/cellranger-6.0.1",
              "GRCh38/BRI-1283/outs/filtered_feature_bc_matrix")

features_df <- file.path(cellranger_dir, "features.tsv.gz") %>%
    read_tsv(col_names = c("gene_id", "gene_name", "phenotype"))

mt_genes <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("^MT-", gene_name)) %>%
    pull(gene_id)

ribo_genes <- features_df %>%
    filter(phenotype == "Gene Expression", 
           grepl("^RPS\\d+|^RPL\\d+", gene_name)) %>%
  pull(gene_id)

data10x <- Read10X(cellranger_dir, gene.column = 1)
```

## Create the Seurat object

```{r, cache=TRUE}
gene_exp <- data10x[["Gene Expression"]]

antibody <- data10x[["Antibody Capture"]] %>%
    .[!grepl("^Hashtag", rownames(.)), ] 

rownames(antibody) <- rownames(antibody) %>%
    sub("_prot$", "", .) %>%
    gsub("_", ".", .)

hashtags <- data10x[["Antibody Capture"]] %>%
    .[grepl("^Hashtag", rownames(.)), ]

rownames(hashtags) <- 
    c("IgG72", "RSQ72", "IgG24", "RSQ24", "Res24", "Res00")

# Create object
bcells <- CreateSeuratObject(counts = gene_exp, project = "bcells")
bcells[["ADT"]] <- CreateAssayObject(counts = antibody)
bcells[["HTO"]] <- CreateAssayObject(counts = hashtags)

# Normalize
bcells <- bcells %>%
  NormalizeData(normalization.method = "LogNormalize", margin = 2) %>%
  NormalizeData(assay = "HTO", normalization.method = "CLR", margin = 2) %>%
  NormalizeData(assay = "ADT", normalization.method = "CLR", margin = 2)

bcells[["percent_mt"]] <- PercentageFeatureSet(bcells, features = mt_genes)
bcells[["percent_ribo"]] <- PercentageFeatureSet(bcells, features = ribo_genes)
```

## QC

Here we use the miQC package to model the percentage of mitochondrial reads and number of genes, in order to identify and remove compromised cells.

```{r, fig.height=2, fig.width=4}
bcells_sce <- bcells %>%
    as.SingleCellExperiment() %>%
    addPerCellQC(subsets = list(mito = mt_genes))

model <- mixtureModel(bcells_sce)

plotFiltering(bcells_sce, model, posterior_cutoff = 0.8) +
  scale_y_continuous(breaks = scales::pretty_breaks(8)) +
  theme_bw()
```

## Remove compromised cells

```{r}
bcells_sce <- filterCells(bcells_sce, model)
cells_keep <- rownames(colData(bcells_sce))
bcells <- subset(bcells, cells = cells_keep)
```

## Demultiplex cells based on HTO

```{r}
bcells <- HTODemux(bcells, assay = "HTO", positive.quantile = 0.99)

table(bcells$HTO_classification.global)

Idents(bcells) <- "HTO_maxID"
```

```{r, echo=FALSE}
stims <- c("Res00", "Res24", "IgG24", "IgG72", "RSQ24", "RSQ72")

stim_colors <- c("Res00" = "grey75", 
                 "Res24" = "grey60", 
                 "IgG24" = "gold1", 
                 "IgG72" = "gold3", 
                 "RSQ24" = "mediumpurple1", 
                 "RSQ72" = "mediumpurple4")

hto_data <- bcells@assays$HTO@data
hto_class_df <- tibble(barcode = colnames(hto_data),
                       maxid = bcells@meta.data$HTO_maxID)

hto_plot_df <- hto_data %>%
    as.data.frame() %>%
    rownames_to_column("hto") %>%
    pivot_longer(-hto, names_to = "barcode") %>%
    left_join(hto_class_df, by = "barcode") %>%
    select(maxid, hto, value) %>%
    mutate(maxid = factor(maxid, levels = stims),
           hto = factor(hto, levels = stims))

ggplot(hto_plot_df, aes(value, maxid, fill = maxid)) +
    geom_density_ridges(show.legend = FALSE, size = .25) +
    facet_wrap(~hto, nrow = 2) +
    scale_fill_manual(values = stim_colors) +
    theme_bw() +
    theme(panel.grid = element_blank()) +
    labs(x = "HTO normalized levels", y = NULL)
```

```{r, echo=FALSE, fig.width=6.5, fig.height=4}
hto_global_df <- 
  tibble(barcode = colnames(hto_data),
         maxid = bcells@meta.data$HTO_maxID,
         classif = bcells@meta.data$HTO_classification.global) %>%
  count(maxid, classif) %>%
  group_by(maxid) %>%
  mutate(p = round(n/sum(n) * 100, 1),
         p = paste0(p, "%"),
         classif = factor(classif, levels = c("Negative", "Doublet", "Singlet")),
         maxid = factor(maxid, levels = stims)) %>%
  ungroup()

ggplot(hto_global_df, aes(classif, n, fill = maxid)) +
  geom_col(show.legend = FALSE, color = "black", size = .5) +
  geom_text(aes(label = p), size = 3, vjust = 0, nudge_y = 10) +
  scale_fill_manual(values = stim_colors) +
  facet_wrap(~maxid, nrow = 1, strip.position = "bottom") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid.major.x = element_blank(),
        panel.spacing = unit(0, "lines"),
        panel.border = element_blank()) +
  labs(x = NULL, y = "Number of droplets")
```
### HTO relative levels in Singlet, Doublet, and Negative droplets

```{r, echo=FALSE, fig.width=10, fig.height=8}
stim_class_df <- tibble(barcode = colnames(bcells@assays$HTO@data),
                        stim = bcells@meta.data$HTO_maxID,
                        classif = bcells@meta.data$HTO_classification.global)

hto_df <- bcells@assays$HTO@data %>%
  as.data.frame() %>%
  rownames_to_column("hto") %>%
  as_tibble() %>%
  pivot_longer(-hto, names_to = "barcode") %>%
  left_join(stim_class_df)

hto_top <- hto_df %>%
  group_by(barcode, classif) %>%
  slice_max(n = 1, order_by = value) %>%
  select(barcode, classif, top_hto = hto)

hto_plot_df <- hto_df %>%
  left_join(hto_top, by = c("barcode", "classif")) %>%
  mutate(top_hto = factor(top_hto, levels = stims),
         stim = factor(stim, stims))

plot_admix <- function(plot_df) {
  ggplot(plot_df, aes(reorder_within(barcode, by = value, within = stim), value)) +
  geom_col(aes(fill = hto), position = "fill", width = 1.01, show.legend = FALSE) +
  scale_fill_manual(values = stim_colors) +
  facet_grid(~stim, scales = "free_x", space = "free") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        strip.text = element_blank()) +
  labs(x = NULL, y = "HTO proportion", fill = "Stim")
}

admix_singlet <- hto_plot_df %>%
  filter(classif == "Singlet") %>%
  plot_admix() +
  labs(title = "Singlets")

admix_doublet <- hto_plot_df %>%
  filter(classif == "Doublet") %>%
  plot_admix() +
  labs(title = "Doublets")

admix_negative <- hto_plot_df %>%
  filter(classif == "Negative") %>%
  plot_admix() +
  labs(title = "Negatives")

plot_grid(admix_singlet, admix_doublet, admix_negative, ncol = 1)
```


## Extract Singlets

```{r}
Idents(bcells) <- "HTO_classification.global"

bcells_singlet <- subset(bcells, idents = "Singlet")

table(bcells_singlet@meta.data$HTO_maxID)[stims]
```

```{r, include=FALSE, eval=FALSE, echo=FALSE}
saveRDS(bcells_singlet, "./scDRS/data/expression/bcells_singlet_seurat.rds")
#bcells_singlet <- readRDS("./scDRS/data/expression/bcells_singlet_seurat.rds")
```



## Feature quantifications

```{r, echo=FALSE, fig.width=10, fig.height=3}
bcells_singlet_meta <- bcells_singlet@meta.data %>%
    rownames_to_column("barcode") %>%
    select(barcode, cell_class = HTO_maxID, 
           n_rna_molecules = nCount_RNA, n_genes = nFeature_RNA, percent_mt, percent_ribo) %>%
    mutate(cell_class = factor(cell_class, levels = stims)) %>%
    pivot_longer(-(barcode:cell_class), names_to = "feature")

ggplot(bcells_singlet_meta, aes(cell_class, value, fill = cell_class)) +
    geom_violin(size = .25) +
    scale_y_continuous(breaks = scales::pretty_breaks(7)) +
    scale_fill_manual(values = stim_colors) +
    facet_wrap(~feature, nrow = 1, scales = "free") +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.title = element_blank(),
          axis.ticks.x = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid.minor = element_blank()) +
    labs(fill = NULL)
```


## PCA

```{r}
bcells_singlet <- bcells_singlet %>%
    FindVariableFeatures(nfeatures = 2000, selection.method = "vst") %>%
    ScaleData(., features = rownames(.)) %>%
    RunPCA(., features = VariableFeatures(.))
```


```{r, echo=FALSE, fig.height=7}
pca_cell_emb <- bcells_singlet@reductions$pca@cell.embeddings %>%
  as.data.frame() %>%
  rownames_to_column("barcode") %>%
  as_tibble() %>%
  left_join(stim_class_df) %>%
  select(barcode, stim, PC_1:PC_4)

pca_plot <- ggplot(pca_cell_emb, aes(PC_1, PC_2)) +
  geom_point(aes(color = stim), size = .5) +
  scale_color_manual(values = stim_colors) +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  guides(color = guide_legend(override.aes = list(size = 2)))

sdev_plot <- tibble(sdev = bcells_singlet@reductions$pca@stdev) %>%
    rownames_to_column("pc") %>%
    mutate(pc = fct_inorder(pc)) %>%
    filter(pc %in% 1:25) %>%
    ggplot(aes(pc, sdev)) +
    geom_line(aes(group = 1)) +
    scale_x_discrete(breaks = c(1, seq(5, 25, 5))) +
    theme_bw() +
    theme(panel.grid.minor = element_blank()) +
    labs(x = "Principal component", y = "Standard deviation")

pca_df <- bcells_singlet@reductions$pca@feature.loadings %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  as_tibble() 

loadings_plot <- pca_df %>%
  select(gene_id, PC_1, PC_2) %>%
  pivot_longer(-gene_id, names_to = "pc") %>%
  mutate(direction = case_when(sign(value) == 1 ~ "+",
                               sign(value) == -1 ~ "-",
                               sign(value) == 0 ~ "0")) %>%
  group_by(pc, direction) %>%
  top_n(15, abs(value)) %>%
  ungroup() %>%
  left_join(select(features_df, gene_id, gene_name)) %>%
  ggplot(aes(value, reorder_within(gene_name, value, pc))) +
  geom_col(aes(fill = direction), show.legend = FALSE) +
  scale_y_discrete(labels = function(x) sub("^([^_]+).*$", "\\1", x)) +
  scale_fill_manual(values = c("+" = "cornflowerblue", "-" = "tomato3")) +
  facet_wrap(~pc, scales = "free") +
  labs(x = "Loading", y = NULL) +
  theme_bw()

plot_grid(plot_grid(pca_plot, sdev_plot, rel_widths = c(1, .7)),
          loadings_plot,
          ncol = 1, rel_heights = c(.7, 1))
```


### Number of genes

```{r, echo=FALSE, fig.width=4, fig.height=3}
pca_plot_df_2 <- pca_cell_emb %>%
  left_join(filter(bcells_singlet_meta, feature == "n_genes") %>% select(barcode, n_genes = value))

ggplot(pca_plot_df_2, aes(PC_1, PC_2)) +
  geom_point(aes(color = n_genes), size = .5) +
  scale_color_viridis_c() +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  guides(color = guide_legend(override.aes = list(size = 2)))

```


## UMAP and clustering

```{r}
bcells_singlet <- bcells_singlet %>%
  RunUMAP(dims = 1:20, verbose = FALSE) %>%
  FindNeighbors(dims = 1:20, verbose = FALSE) %>%
  FindClusters(resolution = 0.5, verbose = FALSE)
```


### HTO classification

```{r, echo = FALSE, fig.height=3, fig.width=10}
umap_df <- as.data.frame(bcells_singlet@reductions$umap@cell.embeddings) %>%
    rownames_to_column("barcode") %>%
    as_tibble() %>%
    mutate(cluster_0.5 = bcells_singlet@meta.data$RNA_snn_res.0.5,
           stim = bcells_singlet@meta.data$HTO_maxID,
           stim = factor(stim, levels = stims),
           percent_mt = bcells_singlet@meta.data$percent_mt,
           percent_ribo = bcells_singlet@meta.data$percent_ribo,
           n_genes = bcells_singlet@meta.data$nFeature_RNA)

umap_stims <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = stim)) +
    geom_point(size = 1) +
    scale_color_manual(values = stim_colors) +
    theme_bw() +
    theme(panel.grid = element_blank(),
          panel.border = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2)))

cluster_labels <- umap_df %>%
  group_by(cluster_0.5) %>%
  summarise_at(vars(UMAP_1, UMAP_2), median) %>%
  ungroup()

umap_clusters <- ggplot(umap_df) +
    geom_point(aes(UMAP_1, UMAP_2, color = cluster_0.5), size = 1) +
    geom_label(data = cluster_labels, 
               aes(UMAP_1, UMAP_2, label = cluster_0.5),
               alpha = .5) +
    scale_color_brewer(palette = "Paired") +
    theme_bw() +
    theme(panel.grid = element_blank(),
          panel.border = element_blank(),
          axis.line = element_blank(),
          axis.ticks = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2)))

plot_grid(umap_stims, umap_clusters, nrow = 1)
```

### Cell cycling

```{r, echo=FALSE, fig.height=5, fig.width=8}
mki_gene_df <- features_df %>%
    filter(gene_name == "MKI67") %>%
    select(gene_id, gene_name)

mki_gene_quant <- bcells_singlet@assays$RNA@data %>% 
    .[mki_gene_df$gene_id, ,drop = FALSE] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(mki_gene_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp")

ki67 <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(mki_gene_quant, by = "barcode") %>%
    ggplot(aes(UMAP_1, UMAP_2, color = gene_exp)) +
    geom_point(size = .75) +
    scale_color_viridis_c() +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "MKI67 RNA expression", color = NULL)

ngenes <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = n_genes)) +
    geom_point(size = .75) +
    scale_color_viridis_c(breaks = scales::pretty_breaks(10)) +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "Number of genes", color = NULL)

mt <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = percent_mt)) +
    geom_point(size = .75) +
    scale_color_viridis_c() +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "% Mitochondrial genes", color = NULL)

ribo <- ggplot(umap_df, aes(UMAP_1, UMAP_2, color = percent_ribo)) +
    geom_point(size = .75) +
    scale_color_viridis_c() +
    theme_minimal() +
    theme(panel.grid = element_blank()) +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    labs(title = "% Ribosomal genes", color = NULL)


plot_grid(ki67, ngenes, mt, ribo, nrow = 2)
```


## Marker genes for Seurat clusters (whole data, res = 0.5)

```{r}
cluster_markers <- 
    FindAllMarkers(bcells_singlet, 
                   only.pos = TRUE,
                   min.pct = 0.05,
                   logfc.threshold = 1) %>%
    as_tibble() %>%
    filter(p_val_adj < 0.05)
```  

## Marker genes per cluster


```{r, echo=FALSE, fig.height=8, fig.width=10}
plot_markers <- function(cluster_df) {
  
  top_markers <- cluster_df %>%
    as_tibble() %>%
    left_join(select(features_df, 1:2), by = c("gene" = "gene_id")) %>%
    group_by(cluster) %>%
    top_n(10, avg_log2FC) %>%
    ungroup() %>%
    select(cluster, gene_name, gene_id = gene, avg_log2FC)
  
  cluster_cell <- tibble(cluster_cell = Idents(bcells_singlet),
                         barcode = names(Idents(bcells_singlet)))
  
  cell_gene_expr <- bcells_singlet@assays$RNA@data %>%
    .[unique(top_markers$gene_id), ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    pivot_longer(-gene_id, names_to = "barcode", values_to = "logexpr")
  
  top_marker_expr <- cell_gene_expr %>%
    left_join(cluster_cell, by = "barcode") %>%
    inner_join(top_markers, by = "gene_id")
  
  top_marker_summary <- top_marker_expr %>%
    group_by(cluster = cluster_cell, gene_name) %>%
    summarise(prop_expr = mean(logexpr > 0),
              scale_expr = mean(logexpr)) %>%
    ungroup() %>%
    left_join(select(top_markers, cluster_top = cluster, gene_name, avg_log2FC)) %>%
    mutate_at(vars(cluster, cluster_top), factor) %>%
    arrange(cluster_top, avg_log2FC) %>%
    mutate(gene_name = fct_inorder(gene_name)) 
  
  
  ggplot(top_marker_summary, aes(cluster, gene_name)) +
    geom_point(aes(size = prop_expr, fill = scale_expr), 
               color = "black", shape = 21) +
    scale_size(range = c(0.1, 4), labels = scales::percent) +
    scale_fill_viridis_c(option = "magma") +
    facet_wrap(~cluster_top, scales = "free", ncol = 3) +
    theme_bw() +
    theme(axis.line = element_blank(),
          axis.text = element_text(size = 8),
          panel.grid = element_line(color = "grey96"),
          panel.border = element_blank(),
          legend.position = "top") +
    labs(x = NULL, y = NULL, 
         fill = "Scaled\nExpression", 
         size = "% of cells") +
    guides(fill = guide_colorbar(barheight = .5))
}

plot_markers(cluster_markers)
```


## Marker genes for each stim condition

```{r}
Idents(bcells_singlet) <- "HTO_maxID"

bcells_markers <- 
    FindAllMarkers(bcells_singlet, 
                   only.pos = TRUE,
                   min.pct = 0.1,
                   logfc.threshold = 1) %>%
    as_tibble()
```  

```{r, echo=FALSE, fig.height=5, fig.width=10}
bcells_markers %>%
  mutate(cluster = factor(cluster, levels = stims)) %>%
  plot_markers()
```



## B cell genes (RNA)

```{r, echo=FALSE, fig.width=9, fig.height=18}
bcell_genes <- 
    c("IGHD", "IGHM", "IGHE", "IGHG1", "IGHG2", "IGHG3", "IGHG4", 
      "CD19", "MS4A1", "CR2", "FCER2", "CD27", "CD38", "CD40", "CD69", "CD86",
      "ITGAX", "TBX21", "CXCR5", "HLA-DRA", 
      "TCL1A", "IL4R",
      "TNFSF13B", "TNFRSF13B", "TNFRSF17", "JCHAIN", "IL6")

bcell_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           gene_name %in% bcell_genes) %>%
    select(gene_id, gene_name)

bcell_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[bcell_genes_df$gene_id, ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(bcell_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp") %>%
    mutate(gene_name = factor(gene_name, levels = bcell_genes))

# shrink outliers
bcell_genes_quant_norm <- bcell_genes_quant %>%
  group_by(gene_id) %>%
  mutate(q01 = quantile(gene_exp, 0.005),
         q99 = quantile(gene_exp, 0.995),
         gene_exp = case_when(gene_exp < q01 ~ q01,
                              gene_exp > q99 ~ q99,
                              TRUE ~ gene_exp)) %>%
  ungroup()

bcell_genes_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(bcell_genes_quant_norm, by = "barcode") %>%
    split(.$gene_name) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = gene_exp)) +
            geom_point(size = .2) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5)) +
            facet_wrap(~gene_name) +
            theme_bw() +
            theme(panel.grid = element_blank(),
                  panel.border = element_blank(),
                  strip.background = element_blank(),
                  axis.line = element_blank(),
                  axis.ticks = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = bcell_genes_plot_list, ncol = 3)
```


## B cell genes (Protein)

```{r, echo=FALSE, fig.height=12, fig.width=9}
bcell_prots <- 
    c("IGHD" = "IgD", "IGHM" = "IgM",
      "CD19" = "CD19", "MS4A1" = "CD20", 
      "CR2" = "CD21", "FCER2" = "CD23",
      "CD27" = "CD27",
      "CD38" = "CD38", "CD40" = "CD40", 
      "CD69" = "CD69", "CD86" = "CD86",
      "ITGAX" = "CD11c", "CXCR5" = "CD185.or.CXCR5",
      "HLA-DRA" = "HLA-DR", 
      "TNFRSF13B" = "CD267.or.TACI",
      "TNFRSF13C" = "CD268.or.BAFF-R")

adt_quants <- bcells_singlet@assays$ADT@data[bcell_prots, ] %>%
    as.data.frame() %>%
    rownames_to_column("ab") %>%
    as_tibble() %>%
    pivot_longer(-ab, names_to = "barcode", values_to = "ab_level") %>%
    mutate(ab = factor(ab, levels = bcell_prots))

# shrink outliers
adt_quants_norm <- adt_quants %>%
  group_by(ab) %>%
  mutate(q01 = quantile(ab_level, 0.005),
         q99 = quantile(ab_level, 0.995),
         ab_level = case_when(ab_level < q01 ~ q01,
                              ab_level > q99 ~ q99,
                              TRUE ~ ab_level)) %>%
  ungroup()

bcell_prots_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(adt_quants_norm, by = "barcode") %>%
    split(.$ab) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = ab_level)) +
            geom_point(size = .2) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5,
                                                         barheight = 3)) +
            facet_wrap(~ab) +
            theme_bw() +
            theme(panel.grid = element_blank(),
                  panel.border = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank(),
                  strip.background = element_blank(),
                  axis.line = element_blank(),
                  axis.ticks = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = bcell_prots_plot_list, ncol = 3)
```


## Lupus genes

```{r, echo=FALSE, fig.height=16, fig.width=10}
sle_genes <- 
    "https://www.nature.com/articles/ng.3434/tables/2" %>%
    read_html() %>%
    html_node("table") %>%
    html_table(header = TRUE, fill = TRUE) %>%
    select(gene = `Likely causal genec`) %>%
    slice(-1) %>%
    separate_rows(gene, sep = ",") %>%
    mutate(gene = trimws(gene)) %>%
    filter(gene != "") %>%
    pull(gene)

#sle_genes %>% write_lines("./scDRS/data/bentham_genes.txt")

sle_genes_df <- features_df %>%
    filter(phenotype == "Gene Expression", 
           gene_name %in% sle_genes) %>%
    select(gene_id, gene_name)

sle_genes_quant <- bcells_singlet@assays$RNA@data %>% 
    .[sle_genes_df$gene_id, ] %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    left_join(sle_genes_df, by = "gene_id") %>%
    pivot_longer(-c("gene_id", "gene_name"), 
                 names_to = "barcode", values_to = "gene_exp") %>%
    group_by(gene_id) %>%
    filter(mean(gene_exp > 0) > .05) %>%
    ungroup()

sle_genes_quant_norm <- sle_genes_quant %>%
  group_by(gene_id) %>%
  mutate(q01 = quantile(gene_exp, 0.005),
         q99 = quantile(gene_exp, 0.995),
         gene_exp = case_when(gene_exp < q01 ~ q01,
                              gene_exp > q99 ~ q99,
                              TRUE ~ gene_exp)) %>%
  ungroup()

sle_genes_plot_list <- umap_df %>%
    select(barcode, UMAP_1, UMAP_2) %>%
    left_join(sle_genes_quant_norm, by = "barcode") %>%
    split(.$gene_name) %>%
    map(~ggplot(data = ., aes(UMAP_1, UMAP_2, color = gene_exp)) +
            geom_point(size = .1) +
            scale_color_viridis_c(labels = function(x) str_pad(x, 3),
                                  guide = guide_colorbar(barwidth = .5)) +
            facet_wrap(~gene_name) +
            theme_minimal() +
            theme(panel.grid = element_blank(),
                  axis.title = element_blank(),
                  axis.text = element_blank()) +
            labs(color = NULL))

plot_grid(plotlist = sle_genes_plot_list, ncol = 4)

```





## Marker genes IgG vs RSQ

### 24 hours

```{r}
bcells_markers_24 <- 
    FindMarkers(bcells_singlet, 
                   ident.1 = "IgG24",
                   ident.2 = "RSQ24",
                   only.pos = FALSE,
                   min.pct = 0.1,
                   logfc.threshold = 0.5) %>%
    rownames_to_column("gene") %>%
    as_tibble() %>%
    select(gene, avg_log2FC, IgG24 = pct.1, RSQ24 = pct.2, p = p_val_adj) %>%
    filter(p < 0.05)
```

```{r, echo=FALSE}
markers_timep_df <- bcells_markers_24 %>%
  mutate(cluster_top = ifelse(avg_log2FC > 0, "IgG24", "RSQ24")) %>%
  left_join(select(features_df, 1:2), by = c("gene" = "gene_id")) %>%
  pivot_longer(IgG24:RSQ24, names_to = "cluster", values_to = "pct")

cell_gene_expr_timep <- bcells_singlet@assays$RNA@data %>%
  .[unique(markers_timep_df$gene), ] %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  as_tibble() %>%
  pivot_longer(-gene_id, names_to = "barcode", values_to = "logexpr")

cluster_cell <- tibble(cluster_cell = Idents(bcells_singlet),
                       barcode = names(Idents(bcells_singlet)))

timep_gene_avg_expr <- cell_gene_expr_timep %>%
  left_join(cluster_cell) %>%
  filter(cluster_cell %in% c("IgG24", "RSQ24")) %>%
  group_by(cluster = cluster_cell, gene_id) %>%
  summarise(scale_expr = mean(logexpr)) %>%
  ungroup() %>%
  left_join(select(features_df, gene_id, gene_name)) %>%
  left_join(markers_timep_df, by = c("gene_id" = "gene", "gene_name", "cluster")) %>%
  arrange(avg_log2FC) %>%
  mutate(gene_name = fct_inorder(gene_name))
```

#### GWAS genes

```{r, echo = FALSE, fig.width=4, fig.height=3}
reported_genes <- read_tsv("./reported_genes.tsv")

de_24_df <- timep_gene_avg_expr %>%
  filter(gene_name %in% unique(reported_genes$gene)) %>%
  filter(! grepl("HLA", gene_name)) %>% 
  mutate_at(vars(cluster, cluster_top),
            ~recode(., "IgG24" = "BCR", "RSQ24" = "TRL7"))


ggplot(de_24_df, aes(cluster, gene_name)) +
   geom_point(aes(size = pct, fill = scale_expr), 
             color = "black", shape = 21) +
  scale_size(range = c(0.25, 5), labels = scales::percent) +
  scale_fill_viridis_c(option = "magma") +
  theme_bw() +
  theme(panel.background = element_blank(),
        plot.margin = margin(t = .5, b = 1, unit = "cm")) +
  labs(x = NULL, y = NULL, fill = "Scaled\nexpression", size = "% of\ncells") +
  guides(fill = guide_colorbar(barwidth = .5, barheight = 3),
         size = guide_legend(keyheight = .5))
```



## MAGMA

Scores taken from the scDRS figshare.

```{r, echo=FALSE, fig.height=6, eval=FALSE}
magma <- "./scDRS/data/MAGMA_v108_GENE_10_ZSTAT_for_scDRS.txt" %>%
  read.table() %>%
  rownames_to_column("gene_name") %>%
  select(gene_name, z = PASS_Lupus) %>%
  drop_na() %>%
  mutate(abs_z = abs(z))

magma_plot_df <- magma %>%
  top_n(50, abs_z) %>%
  arrange(abs_z) %>%
  mutate(gene_name = fct_inorder(gene_name))

ggplot(magma_plot_df, aes(abs_z, gene_name)) +
  geom_point(size = 4, alpha = .5) +
  theme_minimal() +
  theme(panel.grid.minor = element_blank()) +
  labs(x = "Absolute z-score", y = NULL)
```



## scDRS

```{r, echo=FALSE, fig.width=5, fig.height=3, eval=FALSE}
scdrs_scores <- read_tsv("./scDRS/results/Bentham_data_scDRSpaper/SLE.score.gz") %>%
  rename("barcode" = 1)

umap_scdrs_df <- umap_df %>%
  select(1:3) %>%
  left_join(scdrs_scores, by = "barcode")

ggplot(umap_scdrs_df, aes(UMAP_1, UMAP_2, color = norm_score)) +
    geom_point(size = .5) +
    scale_color_gradient2() +
    theme(panel.grid = element_blank(),
          panel.background = element_rect(fill = "grey")) +
    guides(color = guide_legend(override.aes = list(size = 2)))

```





